// @rainbow-me/rainbowkit/dist/chunk-ZESJAA3L.js
import {
  lightTheme
} from "./chunk-4MKWCF32.js";

// src/components/ConnectButton/ConnectButton.tsx
import React42 from "react";

// src/css/sprinkles.css.ts
import { createMapValueFn as _51c72 } from "@vanilla-extract/sprinkles/createUtils";
import { createNormalizeValueFn as _a49f6 } from "@vanilla-extract/sprinkles/createUtils";
import { createSprinkles as _ad221 } from "@vanilla-extract/sprinkles/createRuntimeSprinkles";
var mapResponsiveValue = _51c72({ conditions: { defaultCondition: "smallScreen", conditionNames: ["smallScreen", "largeScreen"], responsiveArray: void 0 } });
var normalizeResponsiveValue = _a49f6({ conditions: { defaultCondition: "smallScreen", conditionNames: ["smallScreen", "largeScreen"], responsiveArray: void 0 } });
var sprinkles = _ad221({ conditions: { defaultCondition: "base", conditionNames: ["base", "hover", "active"], responsiveArray: void 0 }, styles: { background: { values: { accentColor: { conditions: { base: "ju367v94", hover: "ju367v95", active: "ju367v96" }, defaultClass: "ju367v94" }, accentColorForeground: { conditions: { base: "ju367v97", hover: "ju367v98", active: "ju367v99" }, defaultClass: "ju367v97" }, actionButtonBorder: { conditions: { base: "ju367v9a", hover: "ju367v9b", active: "ju367v9c" }, defaultClass: "ju367v9a" }, actionButtonBorderMobile: { conditions: { base: "ju367v9d", hover: "ju367v9e", active: "ju367v9f" }, defaultClass: "ju367v9d" }, actionButtonSecondaryBackground: { conditions: { base: "ju367v9g", hover: "ju367v9h", active: "ju367v9i" }, defaultClass: "ju367v9g" }, closeButton: { conditions: { base: "ju367v9j", hover: "ju367v9k", active: "ju367v9l" }, defaultClass: "ju367v9j" }, closeButtonBackground: { conditions: { base: "ju367v9m", hover: "ju367v9n", active: "ju367v9o" }, defaultClass: "ju367v9m" }, connectButtonBackground: { conditions: { base: "ju367v9p", hover: "ju367v9q", active: "ju367v9r" }, defaultClass: "ju367v9p" }, connectButtonBackgroundError: { conditions: { base: "ju367v9s", hover: "ju367v9t", active: "ju367v9u" }, defaultClass: "ju367v9s" }, connectButtonInnerBackground: { conditions: { base: "ju367v9v", hover: "ju367v9w", active: "ju367v9x" }, defaultClass: "ju367v9v" }, connectButtonText: { conditions: { base: "ju367v9y", hover: "ju367v9z", active: "ju367va0" }, defaultClass: "ju367v9y" }, connectButtonTextError: { conditions: { base: "ju367va1", hover: "ju367va2", active: "ju367va3" }, defaultClass: "ju367va1" }, connectionIndicator: { conditions: { base: "ju367va4", hover: "ju367va5", active: "ju367va6" }, defaultClass: "ju367va4" }, error: { conditions: { base: "ju367va7", hover: "ju367va8", active: "ju367va9" }, defaultClass: "ju367va7" }, generalBorder: { conditions: { base: "ju367vaa", hover: "ju367vab", active: "ju367vac" }, defaultClass: "ju367vaa" }, generalBorderDim: { conditions: { base: "ju367vad", hover: "ju367vae", active: "ju367vaf" }, defaultClass: "ju367vad" }, menuItemBackground: { conditions: { base: "ju367vag", hover: "ju367vah", active: "ju367vai" }, defaultClass: "ju367vag" }, modalBackdrop: { conditions: { base: "ju367vaj", hover: "ju367vak", active: "ju367val" }, defaultClass: "ju367vaj" }, modalBackground: { conditions: { base: "ju367vam", hover: "ju367van", active: "ju367vao" }, defaultClass: "ju367vam" }, modalBorder: { conditions: { base: "ju367vap", hover: "ju367vaq", active: "ju367var" }, defaultClass: "ju367vap" }, modalText: { conditions: { base: "ju367vas", hover: "ju367vat", active: "ju367vau" }, defaultClass: "ju367vas" }, modalTextDim: { conditions: { base: "ju367vav", hover: "ju367vaw", active: "ju367vax" }, defaultClass: "ju367vav" }, modalTextSecondary: { conditions: { base: "ju367vay", hover: "ju367vaz", active: "ju367vb0" }, defaultClass: "ju367vay" }, profileAction: { conditions: { base: "ju367vb1", hover: "ju367vb2", active: "ju367vb3" }, defaultClass: "ju367vb1" }, profileActionHover: { conditions: { base: "ju367vb4", hover: "ju367vb5", active: "ju367vb6" }, defaultClass: "ju367vb4" }, profileForeground: { conditions: { base: "ju367vb7", hover: "ju367vb8", active: "ju367vb9" }, defaultClass: "ju367vb7" }, selectedOptionBorder: { conditions: { base: "ju367vba", hover: "ju367vbb", active: "ju367vbc" }, defaultClass: "ju367vba" }, standby: { conditions: { base: "ju367vbd", hover: "ju367vbe", active: "ju367vbf" }, defaultClass: "ju367vbd" } } }, borderColor: { values: { accentColor: { conditions: { base: "ju367vbg", hover: "ju367vbh", active: "ju367vbi" }, defaultClass: "ju367vbg" }, accentColorForeground: { conditions: { base: "ju367vbj", hover: "ju367vbk", active: "ju367vbl" }, defaultClass: "ju367vbj" }, actionButtonBorder: { conditions: { base: "ju367vbm", hover: "ju367vbn", active: "ju367vbo" }, defaultClass: "ju367vbm" }, actionButtonBorderMobile: { conditions: { base: "ju367vbp", hover: "ju367vbq", active: "ju367vbr" }, defaultClass: "ju367vbp" }, actionButtonSecondaryBackground: { conditions: { base: "ju367vbs", hover: "ju367vbt", active: "ju367vbu" }, defaultClass: "ju367vbs" }, closeButton: { conditions: { base: "ju367vbv", hover: "ju367vbw", active: "ju367vbx" }, defaultClass: "ju367vbv" }, closeButtonBackground: { conditions: { base: "ju367vby", hover: "ju367vbz", active: "ju367vc0" }, defaultClass: "ju367vby" }, connectButtonBackground: { conditions: { base: "ju367vc1", hover: "ju367vc2", active: "ju367vc3" }, defaultClass: "ju367vc1" }, connectButtonBackgroundError: { conditions: { base: "ju367vc4", hover: "ju367vc5", active: "ju367vc6" }, defaultClass: "ju367vc4" }, connectButtonInnerBackground: { conditions: { base: "ju367vc7", hover: "ju367vc8", active: "ju367vc9" }, defaultClass: "ju367vc7" }, connectButtonText: { conditions: { base: "ju367vca", hover: "ju367vcb", active: "ju367vcc" }, defaultClass: "ju367vca" }, connectButtonTextError: { conditions: { base: "ju367vcd", hover: "ju367vce", active: "ju367vcf" }, defaultClass: "ju367vcd" }, connectionIndicator: { conditions: { base: "ju367vcg", hover: "ju367vch", active: "ju367vci" }, defaultClass: "ju367vcg" }, error: { conditions: { base: "ju367vcj", hover: "ju367vck", active: "ju367vcl" }, defaultClass: "ju367vcj" }, generalBorder: { conditions: { base: "ju367vcm", hover: "ju367vcn", active: "ju367vco" }, defaultClass: "ju367vcm" }, generalBorderDim: { conditions: { base: "ju367vcp", hover: "ju367vcq", active: "ju367vcr" }, defaultClass: "ju367vcp" }, menuItemBackground: { conditions: { base: "ju367vcs", hover: "ju367vct", active: "ju367vcu" }, defaultClass: "ju367vcs" }, modalBackdrop: { conditions: { base: "ju367vcv", hover: "ju367vcw", active: "ju367vcx" }, defaultClass: "ju367vcv" }, modalBackground: { conditions: { base: "ju367vcy", hover: "ju367vcz", active: "ju367vd0" }, defaultClass: "ju367vcy" }, modalBorder: { conditions: { base: "ju367vd1", hover: "ju367vd2", active: "ju367vd3" }, defaultClass: "ju367vd1" }, modalText: { conditions: { base: "ju367vd4", hover: "ju367vd5", active: "ju367vd6" }, defaultClass: "ju367vd4" }, modalTextDim: { conditions: { base: "ju367vd7", hover: "ju367vd8", active: "ju367vd9" }, defaultClass: "ju367vd7" }, modalTextSecondary: { conditions: { base: "ju367vda", hover: "ju367vdb", active: "ju367vdc" }, defaultClass: "ju367vda" }, profileAction: { conditions: { base: "ju367vdd", hover: "ju367vde", active: "ju367vdf" }, defaultClass: "ju367vdd" }, profileActionHover: { conditions: { base: "ju367vdg", hover: "ju367vdh", active: "ju367vdi" }, defaultClass: "ju367vdg" }, profileForeground: { conditions: { base: "ju367vdj", hover: "ju367vdk", active: "ju367vdl" }, defaultClass: "ju367vdj" }, selectedOptionBorder: { conditions: { base: "ju367vdm", hover: "ju367vdn", active: "ju367vdo" }, defaultClass: "ju367vdm" }, standby: { conditions: { base: "ju367vdp", hover: "ju367vdq", active: "ju367vdr" }, defaultClass: "ju367vdp" } } }, boxShadow: { values: { connectButton: { conditions: { base: "ju367vds", hover: "ju367vdt", active: "ju367vdu" }, defaultClass: "ju367vds" }, dialog: { conditions: { base: "ju367vdv", hover: "ju367vdw", active: "ju367vdx" }, defaultClass: "ju367vdv" }, profileDetailsAction: { conditions: { base: "ju367vdy", hover: "ju367vdz", active: "ju367ve0" }, defaultClass: "ju367vdy" }, selectedOption: { conditions: { base: "ju367ve1", hover: "ju367ve2", active: "ju367ve3" }, defaultClass: "ju367ve1" }, selectedWallet: { conditions: { base: "ju367ve4", hover: "ju367ve5", active: "ju367ve6" }, defaultClass: "ju367ve4" }, walletLogo: { conditions: { base: "ju367ve7", hover: "ju367ve8", active: "ju367ve9" }, defaultClass: "ju367ve7" } } }, color: { values: { accentColor: { conditions: { base: "ju367vea", hover: "ju367veb", active: "ju367vec" }, defaultClass: "ju367vea" }, accentColorForeground: { conditions: { base: "ju367ved", hover: "ju367vee", active: "ju367vef" }, defaultClass: "ju367ved" }, actionButtonBorder: { conditions: { base: "ju367veg", hover: "ju367veh", active: "ju367vei" }, defaultClass: "ju367veg" }, actionButtonBorderMobile: { conditions: { base: "ju367vej", hover: "ju367vek", active: "ju367vel" }, defaultClass: "ju367vej" }, actionButtonSecondaryBackground: { conditions: { base: "ju367vem", hover: "ju367ven", active: "ju367veo" }, defaultClass: "ju367vem" }, closeButton: { conditions: { base: "ju367vep", hover: "ju367veq", active: "ju367ver" }, defaultClass: "ju367vep" }, closeButtonBackground: { conditions: { base: "ju367ves", hover: "ju367vet", active: "ju367veu" }, defaultClass: "ju367ves" }, connectButtonBackground: { conditions: { base: "ju367vev", hover: "ju367vew", active: "ju367vex" }, defaultClass: "ju367vev" }, connectButtonBackgroundError: { conditions: { base: "ju367vey", hover: "ju367vez", active: "ju367vf0" }, defaultClass: "ju367vey" }, connectButtonInnerBackground: { conditions: { base: "ju367vf1", hover: "ju367vf2", active: "ju367vf3" }, defaultClass: "ju367vf1" }, connectButtonText: { conditions: { base: "ju367vf4", hover: "ju367vf5", active: "ju367vf6" }, defaultClass: "ju367vf4" }, connectButtonTextError: { conditions: { base: "ju367vf7", hover: "ju367vf8", active: "ju367vf9" }, defaultClass: "ju367vf7" }, connectionIndicator: { conditions: { base: "ju367vfa", hover: "ju367vfb", active: "ju367vfc" }, defaultClass: "ju367vfa" }, error: { conditions: { base: "ju367vfd", hover: "ju367vfe", active: "ju367vff" }, defaultClass: "ju367vfd" }, generalBorder: { conditions: { base: "ju367vfg", hover: "ju367vfh", active: "ju367vfi" }, defaultClass: "ju367vfg" }, generalBorderDim: { conditions: { base: "ju367vfj", hover: "ju367vfk", active: "ju367vfl" }, defaultClass: "ju367vfj" }, menuItemBackground: { conditions: { base: "ju367vfm", hover: "ju367vfn", active: "ju367vfo" }, defaultClass: "ju367vfm" }, modalBackdrop: { conditions: { base: "ju367vfp", hover: "ju367vfq", active: "ju367vfr" }, defaultClass: "ju367vfp" }, modalBackground: { conditions: { base: "ju367vfs", hover: "ju367vft", active: "ju367vfu" }, defaultClass: "ju367vfs" }, modalBorder: { conditions: { base: "ju367vfv", hover: "ju367vfw", active: "ju367vfx" }, defaultClass: "ju367vfv" }, modalText: { conditions: { base: "ju367vfy", hover: "ju367vfz", active: "ju367vg0" }, defaultClass: "ju367vfy" }, modalTextDim: { conditions: { base: "ju367vg1", hover: "ju367vg2", active: "ju367vg3" }, defaultClass: "ju367vg1" }, modalTextSecondary: { conditions: { base: "ju367vg4", hover: "ju367vg5", active: "ju367vg6" }, defaultClass: "ju367vg4" }, profileAction: { conditions: { base: "ju367vg7", hover: "ju367vg8", active: "ju367vg9" }, defaultClass: "ju367vg7" }, profileActionHover: { conditions: { base: "ju367vga", hover: "ju367vgb", active: "ju367vgc" }, defaultClass: "ju367vga" }, profileForeground: { conditions: { base: "ju367vgd", hover: "ju367vge", active: "ju367vgf" }, defaultClass: "ju367vgd" }, selectedOptionBorder: { conditions: { base: "ju367vgg", hover: "ju367vgh", active: "ju367vgi" }, defaultClass: "ju367vgg" }, standby: { conditions: { base: "ju367vgj", hover: "ju367vgk", active: "ju367vgl" }, defaultClass: "ju367vgj" } } } } }, { conditions: { defaultCondition: "base", conditionNames: ["base", "hover", "active"], responsiveArray: void 0 }, styles: { transform: { values: { grow: { conditions: { base: "ju367v0", hover: "ju367v1", active: "ju367v2" }, defaultClass: "ju367v0" }, growLg: { conditions: { base: "ju367v3", hover: "ju367v4", active: "ju367v5" }, defaultClass: "ju367v3" }, shrink: { conditions: { base: "ju367v6", hover: "ju367v7", active: "ju367v8" }, defaultClass: "ju367v6" }, shrinkSm: { conditions: { base: "ju367v9", hover: "ju367va", active: "ju367vb" }, defaultClass: "ju367v9" } } }, transition: { values: { "default": { conditions: { base: "ju367vc", hover: "ju367vd", active: "ju367ve" }, defaultClass: "ju367vc" } } } } }, { conditions: { defaultCondition: "smallScreen", conditionNames: ["smallScreen", "largeScreen"], responsiveArray: void 0 }, styles: { alignItems: { values: { "flex-start": { conditions: { smallScreen: "ju367vf", largeScreen: "ju367vg" }, defaultClass: "ju367vf" }, "flex-end": { conditions: { smallScreen: "ju367vh", largeScreen: "ju367vi" }, defaultClass: "ju367vh" }, center: { conditions: { smallScreen: "ju367vj", largeScreen: "ju367vk" }, defaultClass: "ju367vj" } } }, display: { values: { none: { conditions: { smallScreen: "ju367vl", largeScreen: "ju367vm" }, defaultClass: "ju367vl" }, block: { conditions: { smallScreen: "ju367vn", largeScreen: "ju367vo" }, defaultClass: "ju367vn" }, flex: { conditions: { smallScreen: "ju367vp", largeScreen: "ju367vq" }, defaultClass: "ju367vp" } } } } }, { conditions: void 0, styles: { margin: { mappings: ["marginTop", "marginBottom", "marginLeft", "marginRight"] }, marginX: { mappings: ["marginLeft", "marginRight"] }, marginY: { mappings: ["marginTop", "marginBottom"] }, padding: { mappings: ["paddingTop", "paddingBottom", "paddingLeft", "paddingRight"] }, paddingX: { mappings: ["paddingLeft", "paddingRight"] }, paddingY: { mappings: ["paddingTop", "paddingBottom"] }, alignSelf: { values: { "flex-start": { defaultClass: "ju367vr" }, "flex-end": { defaultClass: "ju367vs" }, center: { defaultClass: "ju367vt" } } }, backgroundSize: { values: { cover: { defaultClass: "ju367vu" } } }, borderRadius: { values: { "1": { defaultClass: "ju367vv" }, "6": { defaultClass: "ju367vw" }, "10": { defaultClass: "ju367vx" }, "13": { defaultClass: "ju367vy" }, actionButton: { defaultClass: "ju367vz" }, connectButton: { defaultClass: "ju367v10" }, menuButton: { defaultClass: "ju367v11" }, modal: { defaultClass: "ju367v12" }, modalMobile: { defaultClass: "ju367v13" }, "25%": { defaultClass: "ju367v14" }, full: { defaultClass: "ju367v15" } } }, borderStyle: { values: { solid: { defaultClass: "ju367v16" } } }, borderWidth: { values: { "0": { defaultClass: "ju367v17" }, "1": { defaultClass: "ju367v18" }, "2": { defaultClass: "ju367v19" }, "4": { defaultClass: "ju367v1a" } } }, cursor: { values: { pointer: { defaultClass: "ju367v1b" } } }, flexDirection: { values: { row: { defaultClass: "ju367v1c" }, column: { defaultClass: "ju367v1d" } } }, fontFamily: { values: { body: { defaultClass: "ju367v1e" } } }, fontSize: { values: { "12": { defaultClass: "ju367v1f" }, "13": { defaultClass: "ju367v1g" }, "14": { defaultClass: "ju367v1h" }, "16": { defaultClass: "ju367v1i" }, "18": { defaultClass: "ju367v1j" }, "20": { defaultClass: "ju367v1k" }, "23": { defaultClass: "ju367v1l" } } }, fontWeight: { values: { regular: { defaultClass: "ju367v1m" }, medium: { defaultClass: "ju367v1n" }, semibold: { defaultClass: "ju367v1o" }, bold: { defaultClass: "ju367v1p" }, heavy: { defaultClass: "ju367v1q" } } }, gap: { values: { "0": { defaultClass: "ju367v1r" }, "1": { defaultClass: "ju367v1s" }, "2": { defaultClass: "ju367v1t" }, "4": { defaultClass: "ju367v1u" }, "5": { defaultClass: "ju367v1v" }, "6": { defaultClass: "ju367v1w" }, "8": { defaultClass: "ju367v1x" }, "10": { defaultClass: "ju367v1y" }, "12": { defaultClass: "ju367v1z" }, "14": { defaultClass: "ju367v20" }, "16": { defaultClass: "ju367v21" }, "18": { defaultClass: "ju367v22" }, "20": { defaultClass: "ju367v23" }, "24": { defaultClass: "ju367v24" }, "28": { defaultClass: "ju367v25" }, "32": { defaultClass: "ju367v26" }, "36": { defaultClass: "ju367v27" }, "44": { defaultClass: "ju367v28" }, "64": { defaultClass: "ju367v29" }, "-1": { defaultClass: "ju367v2a" } } }, height: { values: { "1": { defaultClass: "ju367v2b" }, "2": { defaultClass: "ju367v2c" }, "4": { defaultClass: "ju367v2d" }, "8": { defaultClass: "ju367v2e" }, "12": { defaultClass: "ju367v2f" }, "20": { defaultClass: "ju367v2g" }, "24": { defaultClass: "ju367v2h" }, "28": { defaultClass: "ju367v2i" }, "30": { defaultClass: "ju367v2j" }, "32": { defaultClass: "ju367v2k" }, "34": { defaultClass: "ju367v2l" }, "36": { defaultClass: "ju367v2m" }, "40": { defaultClass: "ju367v2n" }, "48": { defaultClass: "ju367v2o" }, "54": { defaultClass: "ju367v2p" }, "60": { defaultClass: "ju367v2q" }, full: { defaultClass: "ju367v2r" }, max: { defaultClass: "ju367v2s" } } }, justifyContent: { values: { "flex-start": { defaultClass: "ju367v2t" }, "flex-end": { defaultClass: "ju367v2u" }, center: { defaultClass: "ju367v2v" }, "space-between": { defaultClass: "ju367v2w" } } }, textAlign: { values: { left: { defaultClass: "ju367v2x" }, center: { defaultClass: "ju367v2y" }, inherit: { defaultClass: "ju367v2z" } } }, marginBottom: { values: { "0": { defaultClass: "ju367v30" }, "1": { defaultClass: "ju367v31" }, "2": { defaultClass: "ju367v32" }, "4": { defaultClass: "ju367v33" }, "5": { defaultClass: "ju367v34" }, "6": { defaultClass: "ju367v35" }, "8": { defaultClass: "ju367v36" }, "10": { defaultClass: "ju367v37" }, "12": { defaultClass: "ju367v38" }, "14": { defaultClass: "ju367v39" }, "16": { defaultClass: "ju367v3a" }, "18": { defaultClass: "ju367v3b" }, "20": { defaultClass: "ju367v3c" }, "24": { defaultClass: "ju367v3d" }, "28": { defaultClass: "ju367v3e" }, "32": { defaultClass: "ju367v3f" }, "36": { defaultClass: "ju367v3g" }, "44": { defaultClass: "ju367v3h" }, "64": { defaultClass: "ju367v3i" }, "-1": { defaultClass: "ju367v3j" } } }, marginLeft: { values: { "0": { defaultClass: "ju367v3k" }, "1": { defaultClass: "ju367v3l" }, "2": { defaultClass: "ju367v3m" }, "4": { defaultClass: "ju367v3n" }, "5": { defaultClass: "ju367v3o" }, "6": { defaultClass: "ju367v3p" }, "8": { defaultClass: "ju367v3q" }, "10": { defaultClass: "ju367v3r" }, "12": { defaultClass: "ju367v3s" }, "14": { defaultClass: "ju367v3t" }, "16": { defaultClass: "ju367v3u" }, "18": { defaultClass: "ju367v3v" }, "20": { defaultClass: "ju367v3w" }, "24": { defaultClass: "ju367v3x" }, "28": { defaultClass: "ju367v3y" }, "32": { defaultClass: "ju367v3z" }, "36": { defaultClass: "ju367v40" }, "44": { defaultClass: "ju367v41" }, "64": { defaultClass: "ju367v42" }, "-1": { defaultClass: "ju367v43" } } }, marginRight: { values: { "0": { defaultClass: "ju367v44" }, "1": { defaultClass: "ju367v45" }, "2": { defaultClass: "ju367v46" }, "4": { defaultClass: "ju367v47" }, "5": { defaultClass: "ju367v48" }, "6": { defaultClass: "ju367v49" }, "8": { defaultClass: "ju367v4a" }, "10": { defaultClass: "ju367v4b" }, "12": { defaultClass: "ju367v4c" }, "14": { defaultClass: "ju367v4d" }, "16": { defaultClass: "ju367v4e" }, "18": { defaultClass: "ju367v4f" }, "20": { defaultClass: "ju367v4g" }, "24": { defaultClass: "ju367v4h" }, "28": { defaultClass: "ju367v4i" }, "32": { defaultClass: "ju367v4j" }, "36": { defaultClass: "ju367v4k" }, "44": { defaultClass: "ju367v4l" }, "64": { defaultClass: "ju367v4m" }, "-1": { defaultClass: "ju367v4n" } } }, marginTop: { values: { "0": { defaultClass: "ju367v4o" }, "1": { defaultClass: "ju367v4p" }, "2": { defaultClass: "ju367v4q" }, "4": { defaultClass: "ju367v4r" }, "5": { defaultClass: "ju367v4s" }, "6": { defaultClass: "ju367v4t" }, "8": { defaultClass: "ju367v4u" }, "10": { defaultClass: "ju367v4v" }, "12": { defaultClass: "ju367v4w" }, "14": { defaultClass: "ju367v4x" }, "16": { defaultClass: "ju367v4y" }, "18": { defaultClass: "ju367v4z" }, "20": { defaultClass: "ju367v50" }, "24": { defaultClass: "ju367v51" }, "28": { defaultClass: "ju367v52" }, "32": { defaultClass: "ju367v53" }, "36": { defaultClass: "ju367v54" }, "44": { defaultClass: "ju367v55" }, "64": { defaultClass: "ju367v56" }, "-1": { defaultClass: "ju367v57" } } }, maxWidth: { values: { "1": { defaultClass: "ju367v58" }, "2": { defaultClass: "ju367v59" }, "4": { defaultClass: "ju367v5a" }, "8": { defaultClass: "ju367v5b" }, "12": { defaultClass: "ju367v5c" }, "20": { defaultClass: "ju367v5d" }, "24": { defaultClass: "ju367v5e" }, "28": { defaultClass: "ju367v5f" }, "30": { defaultClass: "ju367v5g" }, "32": { defaultClass: "ju367v5h" }, "34": { defaultClass: "ju367v5i" }, "36": { defaultClass: "ju367v5j" }, "40": { defaultClass: "ju367v5k" }, "48": { defaultClass: "ju367v5l" }, "54": { defaultClass: "ju367v5m" }, "60": { defaultClass: "ju367v5n" }, full: { defaultClass: "ju367v5o" }, max: { defaultClass: "ju367v5p" } } }, minWidth: { values: { "1": { defaultClass: "ju367v5q" }, "2": { defaultClass: "ju367v5r" }, "4": { defaultClass: "ju367v5s" }, "8": { defaultClass: "ju367v5t" }, "12": { defaultClass: "ju367v5u" }, "20": { defaultClass: "ju367v5v" }, "24": { defaultClass: "ju367v5w" }, "28": { defaultClass: "ju367v5x" }, "30": { defaultClass: "ju367v5y" }, "32": { defaultClass: "ju367v5z" }, "34": { defaultClass: "ju367v60" }, "36": { defaultClass: "ju367v61" }, "40": { defaultClass: "ju367v62" }, "48": { defaultClass: "ju367v63" }, "54": { defaultClass: "ju367v64" }, "60": { defaultClass: "ju367v65" }, full: { defaultClass: "ju367v66" }, max: { defaultClass: "ju367v67" } } }, overflow: { values: { hidden: { defaultClass: "ju367v68" } } }, paddingBottom: { values: { "0": { defaultClass: "ju367v69" }, "1": { defaultClass: "ju367v6a" }, "2": { defaultClass: "ju367v6b" }, "4": { defaultClass: "ju367v6c" }, "5": { defaultClass: "ju367v6d" }, "6": { defaultClass: "ju367v6e" }, "8": { defaultClass: "ju367v6f" }, "10": { defaultClass: "ju367v6g" }, "12": { defaultClass: "ju367v6h" }, "14": { defaultClass: "ju367v6i" }, "16": { defaultClass: "ju367v6j" }, "18": { defaultClass: "ju367v6k" }, "20": { defaultClass: "ju367v6l" }, "24": { defaultClass: "ju367v6m" }, "28": { defaultClass: "ju367v6n" }, "32": { defaultClass: "ju367v6o" }, "36": { defaultClass: "ju367v6p" }, "44": { defaultClass: "ju367v6q" }, "64": { defaultClass: "ju367v6r" }, "-1": { defaultClass: "ju367v6s" } } }, paddingLeft: { values: { "0": { defaultClass: "ju367v6t" }, "1": { defaultClass: "ju367v6u" }, "2": { defaultClass: "ju367v6v" }, "4": { defaultClass: "ju367v6w" }, "5": { defaultClass: "ju367v6x" }, "6": { defaultClass: "ju367v6y" }, "8": { defaultClass: "ju367v6z" }, "10": { defaultClass: "ju367v70" }, "12": { defaultClass: "ju367v71" }, "14": { defaultClass: "ju367v72" }, "16": { defaultClass: "ju367v73" }, "18": { defaultClass: "ju367v74" }, "20": { defaultClass: "ju367v75" }, "24": { defaultClass: "ju367v76" }, "28": { defaultClass: "ju367v77" }, "32": { defaultClass: "ju367v78" }, "36": { defaultClass: "ju367v79" }, "44": { defaultClass: "ju367v7a" }, "64": { defaultClass: "ju367v7b" }, "-1": { defaultClass: "ju367v7c" } } }, paddingRight: { values: { "0": { defaultClass: "ju367v7d" }, "1": { defaultClass: "ju367v7e" }, "2": { defaultClass: "ju367v7f" }, "4": { defaultClass: "ju367v7g" }, "5": { defaultClass: "ju367v7h" }, "6": { defaultClass: "ju367v7i" }, "8": { defaultClass: "ju367v7j" }, "10": { defaultClass: "ju367v7k" }, "12": { defaultClass: "ju367v7l" }, "14": { defaultClass: "ju367v7m" }, "16": { defaultClass: "ju367v7n" }, "18": { defaultClass: "ju367v7o" }, "20": { defaultClass: "ju367v7p" }, "24": { defaultClass: "ju367v7q" }, "28": { defaultClass: "ju367v7r" }, "32": { defaultClass: "ju367v7s" }, "36": { defaultClass: "ju367v7t" }, "44": { defaultClass: "ju367v7u" }, "64": { defaultClass: "ju367v7v" }, "-1": { defaultClass: "ju367v7w" } } }, paddingTop: { values: { "0": { defaultClass: "ju367v7x" }, "1": { defaultClass: "ju367v7y" }, "2": { defaultClass: "ju367v7z" }, "4": { defaultClass: "ju367v80" }, "5": { defaultClass: "ju367v81" }, "6": { defaultClass: "ju367v82" }, "8": { defaultClass: "ju367v83" }, "10": { defaultClass: "ju367v84" }, "12": { defaultClass: "ju367v85" }, "14": { defaultClass: "ju367v86" }, "16": { defaultClass: "ju367v87" }, "18": { defaultClass: "ju367v88" }, "20": { defaultClass: "ju367v89" }, "24": { defaultClass: "ju367v8a" }, "28": { defaultClass: "ju367v8b" }, "32": { defaultClass: "ju367v8c" }, "36": { defaultClass: "ju367v8d" }, "44": { defaultClass: "ju367v8e" }, "64": { defaultClass: "ju367v8f" }, "-1": { defaultClass: "ju367v8g" } } }, position: { values: { absolute: { defaultClass: "ju367v8h" }, fixed: { defaultClass: "ju367v8i" }, relative: { defaultClass: "ju367v8j" } } }, right: { values: { "0": { defaultClass: "ju367v8k" } } }, userSelect: { values: { none: { defaultClass: "ju367v8l" } } }, width: { values: { "1": { defaultClass: "ju367v8m" }, "2": { defaultClass: "ju367v8n" }, "4": { defaultClass: "ju367v8o" }, "8": { defaultClass: "ju367v8p" }, "12": { defaultClass: "ju367v8q" }, "20": { defaultClass: "ju367v8r" }, "24": { defaultClass: "ju367v8s" }, "28": { defaultClass: "ju367v8t" }, "30": { defaultClass: "ju367v8u" }, "32": { defaultClass: "ju367v8v" }, "34": { defaultClass: "ju367v8w" }, "36": { defaultClass: "ju367v8x" }, "40": { defaultClass: "ju367v8y" }, "48": { defaultClass: "ju367v8z" }, "54": { defaultClass: "ju367v90" }, "60": { defaultClass: "ju367v91" }, full: { defaultClass: "ju367v92" }, max: { defaultClass: "ju367v93" } } } } });
var themeVars = { colors: { accentColor: "var(--rk-colors-accentColor)", accentColorForeground: "var(--rk-colors-accentColorForeground)", actionButtonBorder: "var(--rk-colors-actionButtonBorder)", actionButtonBorderMobile: "var(--rk-colors-actionButtonBorderMobile)", actionButtonSecondaryBackground: "var(--rk-colors-actionButtonSecondaryBackground)", closeButton: "var(--rk-colors-closeButton)", closeButtonBackground: "var(--rk-colors-closeButtonBackground)", connectButtonBackground: "var(--rk-colors-connectButtonBackground)", connectButtonBackgroundError: "var(--rk-colors-connectButtonBackgroundError)", connectButtonInnerBackground: "var(--rk-colors-connectButtonInnerBackground)", connectButtonText: "var(--rk-colors-connectButtonText)", connectButtonTextError: "var(--rk-colors-connectButtonTextError)", connectionIndicator: "var(--rk-colors-connectionIndicator)", error: "var(--rk-colors-error)", generalBorder: "var(--rk-colors-generalBorder)", generalBorderDim: "var(--rk-colors-generalBorderDim)", menuItemBackground: "var(--rk-colors-menuItemBackground)", modalBackdrop: "var(--rk-colors-modalBackdrop)", modalBackground: "var(--rk-colors-modalBackground)", modalBorder: "var(--rk-colors-modalBorder)", modalText: "var(--rk-colors-modalText)", modalTextDim: "var(--rk-colors-modalTextDim)", modalTextSecondary: "var(--rk-colors-modalTextSecondary)", profileAction: "var(--rk-colors-profileAction)", profileActionHover: "var(--rk-colors-profileActionHover)", profileForeground: "var(--rk-colors-profileForeground)", selectedOptionBorder: "var(--rk-colors-selectedOptionBorder)", standby: "var(--rk-colors-standby)" }, fonts: { body: "var(--rk-fonts-body)" }, radii: { actionButton: "var(--rk-radii-actionButton)", connectButton: "var(--rk-radii-connectButton)", menuButton: "var(--rk-radii-menuButton)", modal: "var(--rk-radii-modal)", modalMobile: "var(--rk-radii-modalMobile)" }, shadows: { connectButton: "var(--rk-shadows-connectButton)", dialog: "var(--rk-shadows-dialog)", profileDetailsAction: "var(--rk-shadows-profileDetailsAction)", selectedOption: "var(--rk-shadows-selectedOption)", selectedWallet: "var(--rk-shadows-selectedWallet)", walletLogo: "var(--rk-shadows-walletLogo)" } };

// src/utils/isMobile.ts
import {
  detect
} from "detect-browser";
function detectEnv(userAgent) {
  return detect(userAgent);
}
function isAndroid() {
  const os = detectOS();
  return os ? os.toLowerCase().includes("android") : false;
}
function isIOS() {
  const os = detectOS();
  return os ? os.toLowerCase().includes("ios") || os.toLowerCase().includes("mac") && navigator.maxTouchPoints > 1 : false;
}
function detectOS() {
  const env = detectEnv();
  return (env == null ? void 0 : env.os) ? env.os : void 0;
}
function isMobile() {
  const os = detectOS();
  return os ? isAndroid() || isIOS() : false;
}

// src/components/AsyncImage/AsyncImage.tsx
import React2, { useReducer as useReducer2 } from "react";

// src/components/Box/Box.ts
import clsx2 from "clsx";
import * as React from "react";

// src/css/atoms.ts
import clsx from "clsx";

// src/css/reset.css.ts
var base = "iekbcc0";
var element = { a: "iekbcca", blockquote: "iekbcc2", button: "iekbcc9", input: "iekbcc8 iekbcc5 iekbcc4", mark: "iekbcc6", ol: "iekbcc1", q: "iekbcc2", select: "iekbcc7 iekbcc5 iekbcc4", table: "iekbcc3", textarea: "iekbcc5 iekbcc4", ul: "iekbcc1" };

// src/css/atoms.ts
var atoms = ({ reset, ...rest }) => {
  if (!reset)
    return sprinkles(rest);
  const elementReset = element[reset];
  const sprinklesClasses = sprinkles(rest);
  return clsx(base, elementReset, sprinklesClasses);
};

// src/components/Box/Box.ts
var Box = React.forwardRef(({ as = "div", className, ...props }, ref) => {
  const atomProps = {};
  const nativeProps = {};
  for (const key in props) {
    if (sprinkles.properties.has(key)) {
      atomProps[key] = props[key];
    } else {
      nativeProps[key] = props[key];
    }
  }
  const atomicClasses = atoms({
    reset: typeof as === "string" ? as : "div",
    ...atomProps
  });
  return React.createElement(as, {
    className: clsx2(atomicClasses, className),
    ...nativeProps,
    ref
  });
});
Box.displayName = "Box";

// src/components/AsyncImage/useAsyncImage.ts
import { useEffect, useReducer } from "react";
var cachedUrls = /* @__PURE__ */ new Map();
var cachedRequestPromises = /* @__PURE__ */ new Map();
async function loadAsyncImage(asyncImage) {
  const cachedRequestPromise = cachedRequestPromises.get(asyncImage);
  if (cachedRequestPromise) {
    return cachedRequestPromise;
  }
  const load = async () => asyncImage().then(async (url) => {
    cachedUrls.set(asyncImage, url);
    return url;
  });
  const requestPromise = load().catch((_err) => {
    return load().catch((_err2) => {
      cachedRequestPromises.delete(asyncImage);
    });
  });
  cachedRequestPromises.set(asyncImage, requestPromise);
  return requestPromise;
}
async function loadImages(...urls) {
  return await Promise.all(urls.map((url) => typeof url === "function" ? loadAsyncImage(url) : url));
}
function useForceUpdate() {
  const [, forceUpdate] = useReducer((x) => x + 1, 0);
  return forceUpdate;
}
function useAsyncImage(url) {
  const cachedUrl = typeof url === "function" ? cachedUrls.get(url) : void 0;
  const forceUpdate = useForceUpdate();
  useEffect(() => {
    if (typeof url === "function" && !cachedUrl) {
      loadAsyncImage(url).then(forceUpdate);
    }
  }, [url, cachedUrl, forceUpdate]);
  return typeof url === "function" ? cachedUrl : url;
}

// src/components/AsyncImage/AsyncImage.tsx
function AsyncImage({
  alt,
  background,
  borderColor,
  borderRadius,
  boxShadow,
  height,
  src: srcProp,
  width
}) {
  const src5 = useAsyncImage(srcProp);
  const isRemoteImage = src5 && /^http/.test(src5);
  const [isRemoteImageLoaded, setRemoteImageLoaded] = useReducer2(() => true, false);
  return /* @__PURE__ */ React2.createElement(Box, {
    "aria-label": alt,
    borderRadius,
    boxShadow,
    height: typeof height === "string" ? height : void 0,
    overflow: "hidden",
    position: "relative",
    style: {
      background,
      height: typeof height === "number" ? height : void 0,
      width: typeof width === "number" ? width : void 0
    },
    width: typeof width === "string" ? width : void 0
  }, /* @__PURE__ */ React2.createElement(Box, {
    ...isRemoteImage ? {
      "aria-hidden": true,
      "as": "img",
      "onLoad": setRemoteImageLoaded,
      "src": src5
    } : {
      backgroundSize: "cover"
    },
    height: "full",
    position: "absolute",
    style: {
      touchCallout: "none",
      transition: "opacity .15s linear",
      userSelect: "none",
      ...isRemoteImage ? {
        opacity: isRemoteImageLoaded ? 1 : 0
      } : {
        backgroundImage: src5 ? `url(${src5})` : void 0,
        backgroundRepeat: "no-repeat",
        opacity: src5 ? 1 : 0
      }
    },
    width: "full"
  }), borderColor ? /* @__PURE__ */ React2.createElement(Box, {
    ...typeof borderColor === "object" && "custom" in borderColor ? { style: { borderColor: borderColor.custom } } : { borderColor },
    borderRadius,
    borderStyle: "solid",
    borderWidth: "1",
    height: "full",
    position: "relative",
    width: "full"
  }) : null);
}

// src/components/Avatar/Avatar.tsx
import React4, { useMemo as useMemo2 } from "react";

// src/components/Icons/Spinner.tsx
import React3, { useMemo } from "react";

// src/components/Icons/Icons.css.ts
var SpinnerIconClassName = "_1luule42";
var SpinnerIconPathClassName = "_1luule43";

// src/components/Icons/Spinner.tsx
var useRandomId = (prefix) => useMemo(() => `${prefix}_${Math.round(Math.random() * 1e9)}`, [prefix]);
var SpinnerIcon = ({
  height = 21,
  width = 21
}) => {
  const id = useRandomId("spinner");
  return /* @__PURE__ */ React3.createElement("svg", {
    className: SpinnerIconClassName,
    fill: "none",
    height,
    viewBox: "0 0 21 21",
    width,
    xmlns: "http://www.w3.org/2000/svg"
  }, /* @__PURE__ */ React3.createElement("clipPath", {
    id
  }, /* @__PURE__ */ React3.createElement("path", {
    d: "M10.5 3C6.35786 3 3 6.35786 3 10.5C3 14.6421 6.35786 18 10.5 18C11.3284 18 12 18.6716 12 19.5C12 20.3284 11.3284 21 10.5 21C4.70101 21 0 16.299 0 10.5C0 4.70101 4.70101 0 10.5 0C16.299 0 21 4.70101 21 10.5C21 11.3284 20.3284 12 19.5 12C18.6716 12 18 11.3284 18 10.5C18 6.35786 14.6421 3 10.5 3Z"
  })), /* @__PURE__ */ React3.createElement("foreignObject", {
    clipPath: `url(#${id})`,
    height: "21",
    width: "21",
    x: "0",
    y: "0"
  }, /* @__PURE__ */ React3.createElement("div", {
    className: SpinnerIconPathClassName
  })));
};

// src/components/Avatar/emojiAvatarForAddress.ts
var colors = [
  "#FC5C54",
  "#FFD95A",
  "#E95D72",
  "#6A87C8",
  "#5FD0F3",
  "#75C06B",
  "#FFDD86",
  "#5FC6D4",
  "#FF949A",
  "#FF8024",
  "#9BA1A4",
  "#EC66FF",
  "#FF8CBC",
  "#FF9A23",
  "#C5DADB",
  "#A8CE63",
  "#71ABFF",
  "#FFE279",
  "#B6B1B6",
  "#FF6780",
  "#A575FF",
  "#4D82FF",
  "#FFB35A"
];
var avatars = [
  { color: colors[0], emoji: "\u{1F336}" },
  { color: colors[1], emoji: "\u{1F911}" },
  { color: colors[2], emoji: "\u{1F419}" },
  { color: colors[3], emoji: "\u{1FAD0}" },
  { color: colors[4], emoji: "\u{1F433}" },
  { color: colors[0], emoji: "\u{1F936}" },
  { color: colors[5], emoji: "\u{1F332}" },
  { color: colors[6], emoji: "\u{1F31E}" },
  { color: colors[7], emoji: "\u{1F412}" },
  { color: colors[8], emoji: "\u{1F435}" },
  { color: colors[9], emoji: "\u{1F98A}" },
  { color: colors[10], emoji: "\u{1F43C}" },
  { color: colors[11], emoji: "\u{1F984}" },
  { color: colors[12], emoji: "\u{1F437}" },
  { color: colors[13], emoji: "\u{1F427}" },
  { color: colors[8], emoji: "\u{1F9A9}" },
  { color: colors[14], emoji: "\u{1F47D}" },
  { color: colors[0], emoji: "\u{1F388}" },
  { color: colors[8], emoji: "\u{1F349}" },
  { color: colors[1], emoji: "\u{1F389}" },
  { color: colors[15], emoji: "\u{1F432}" },
  { color: colors[16], emoji: "\u{1F30E}" },
  { color: colors[17], emoji: "\u{1F34A}" },
  { color: colors[18], emoji: "\u{1F42D}" },
  { color: colors[19], emoji: "\u{1F363}" },
  { color: colors[1], emoji: "\u{1F425}" },
  { color: colors[20], emoji: "\u{1F47E}" },
  { color: colors[15], emoji: "\u{1F966}" },
  { color: colors[0], emoji: "\u{1F479}" },
  { color: colors[17], emoji: "\u{1F640}" },
  { color: colors[4], emoji: "\u26F1" },
  { color: colors[21], emoji: "\u26F5\uFE0F" },
  { color: colors[17], emoji: "\u{1F973}" },
  { color: colors[8], emoji: "\u{1F92F}" },
  { color: colors[22], emoji: "\u{1F920}" }
];
function hashCode(text) {
  let hash = 0;
  if (text.length === 0)
    return hash;
  for (let i = 0; i < text.length; i++) {
    const chr = text.charCodeAt(i);
    hash = (hash << 5) - hash + chr;
    hash |= 0;
  }
  return hash;
}
function emojiAvatarForAddress(address) {
  const resolvedAddress = typeof address === "string" ? address : "";
  const avatarIndex = Math.abs(hashCode(resolvedAddress.toLowerCase()) % avatars.length);
  return avatars[avatarIndex != null ? avatarIndex : 0];
}

// src/components/Avatar/Avatar.tsx
function Avatar({ address, imageUrl, loading, size }) {
  const { color: backgroundColor, emoji } = useMemo2(() => emojiAvatarForAddress(address), [address]);
  return /* @__PURE__ */ React4.createElement(Box, {
    "aria-hidden": true,
    borderRadius: "full",
    overflow: "hidden",
    position: "relative",
    style: {
      height: `${size}px`,
      width: `${size}px`
    },
    userSelect: "none"
  }, /* @__PURE__ */ React4.createElement(Box, {
    alignItems: "center",
    borderRadius: "full",
    display: "flex",
    justifyContent: "center",
    overflow: "hidden",
    position: "absolute",
    style: {
      backgroundColor,
      fontSize: `${Math.round(size * 0.55)}px`,
      height: `${size}px`,
      transform: loading ? "scale(0.72)" : void 0,
      transition: ".25s ease",
      transitionDelay: loading ? void 0 : ".1s",
      width: `${size}px`,
      willChange: "transform"
    },
    userSelect: "none"
  }, imageUrl ? /* @__PURE__ */ React4.createElement(Box, {
    backgroundSize: "cover",
    borderRadius: "full",
    height: "full",
    position: "absolute",
    style: { backgroundImage: `url(${imageUrl})` },
    width: "full"
  }) : null, emoji), typeof loading === "boolean" && /* @__PURE__ */ React4.createElement(Box, {
    color: "accentColor",
    display: "flex",
    height: "full",
    position: "absolute",
    style: {
      opacity: loading ? 1 : 0,
      transition: loading ? "0.6s ease" : "0.2s ease",
      transitionDelay: loading ? ".05s" : void 0
    },
    width: "full"
  }, /* @__PURE__ */ React4.createElement(SpinnerIcon, {
    height: "100%",
    width: "100%"
  })));
}

// src/components/Icons/Dropdown.tsx
import React5 from "react";
var DropdownIcon = () => /* @__PURE__ */ React5.createElement("svg", {
  fill: "none",
  height: "7",
  width: "14",
  xmlns: "http://www.w3.org/2000/svg"
}, /* @__PURE__ */ React5.createElement("path", {
  d: "M12.75 1.54001L8.51647 5.0038C7.77974 5.60658 6.72026 5.60658 5.98352 5.0038L1.75 1.54001",
  stroke: "currentColor",
  strokeLinecap: "round",
  strokeLinejoin: "round",
  strokeWidth: "2.5",
  xmlns: "http://www.w3.org/2000/svg"
}));

// src/components/RainbowKitProvider/RainbowKitChainContext.ts
import { createContext, useContext, useMemo as useMemo3 } from "react";
var RainbowKitChainContext = createContext([]);
var useRainbowKitChains = () => useContext(RainbowKitChainContext);
var useRainbowKitChainsById = () => {
  const rainbowkitChains = useRainbowKitChains();
  return useMemo3(() => {
    const rainbowkitChainsById = {};
    rainbowkitChains.forEach((rkChain) => {
      rainbowkitChainsById[rkChain.id] = rkChain;
    });
    return rainbowkitChainsById;
  }, [rainbowkitChains]);
};

// src/components/ConnectButton/ConnectButtonRenderer.tsx
import React41, {
  useCallback as useCallback7,
  useContext as useContext10,
  useEffect as useEffect13,
  useState as useState9
} from "react";
import {
  useAccount as useAccount4,
  useBalance,
  useDisconnect,
  useEnsAvatar,
  useEnsName,
  useNetwork as useNetwork5
} from "wagmi";

// src/hooks/useIsMounted.ts
import { useEffect as useEffect2, useReducer as useReducer3 } from "react";
var useIsMounted = () => {
  const [mounted, setMounted] = useReducer3(() => true, false);
  useEffect2(setMounted, [setMounted]);
  return mounted;
};

// src/transactions/useRecentTransactions.ts
import { useEffect as useEffect4, useState as useState2 } from "react";

// src/hooks/useAddress.ts
import { useAccount } from "wagmi";
function useAddress() {
  var _a;
  const { data: accountData } = useAccount();
  return (_a = accountData == null ? void 0 : accountData.address) != null ? _a : null;
}

// src/hooks/useChainId.ts
import { useNetwork } from "wagmi";
function useChainId() {
  var _a;
  const { activeChain } = useNetwork();
  return (_a = activeChain == null ? void 0 : activeChain.id) != null ? _a : null;
}

// src/transactions/TransactionStoreContext.tsx
import React6, { createContext as createContext2, useContext as useContext2, useEffect as useEffect3, useState } from "react";
import { useProvider } from "wagmi";

// src/transactions/transactionStore.ts
var storageKey = "rk-transactions";
function safeParseJsonData(string) {
  try {
    const value = string ? JSON.parse(string) : {};
    return typeof value === "object" ? value : {};
  } catch (err) {
    return {};
  }
}
function loadData() {
  return safeParseJsonData(typeof localStorage !== "undefined" ? localStorage.getItem(storageKey) : null);
}
var transactionHashRegex = /^0x([A-Fa-f0-9]{64})$/;
function validateTransaction(transaction) {
  const errors = [];
  if (!transactionHashRegex.test(transaction.hash)) {
    errors.push("Invalid transaction hash");
  }
  if (typeof transaction.description !== "string") {
    errors.push("Transaction must have a description");
  }
  if (typeof transaction.confirmations !== "undefined" && (!Number.isInteger(transaction.confirmations) || transaction.confirmations < 1)) {
    errors.push("Transaction confirmations must be a positiver integer");
  }
  return errors;
}
function createTransactionStore({
  provider: initialProvider
}) {
  let data = loadData();
  let provider = initialProvider;
  const listeners = /* @__PURE__ */ new Set();
  const transactionRequestCache = /* @__PURE__ */ new Map();
  function setProvider(newProvider) {
    provider = newProvider;
  }
  function getTransactions(account, chainId) {
    var _a, _b;
    return (_b = (_a = data[account]) == null ? void 0 : _a[chainId]) != null ? _b : [];
  }
  function addTransaction(account, chainId, transaction) {
    const errors = validateTransaction(transaction);
    if (errors.length > 0) {
      throw new Error(["Unable to add transaction", ...errors].join("\n"));
    }
    updateTransactions(account, chainId, (transactions) => {
      return [
        { ...transaction, status: "pending" },
        ...transactions.filter(({ hash }) => {
          return hash !== transaction.hash;
        })
      ];
    });
  }
  function clearTransactions(account, chainId) {
    updateTransactions(account, chainId, () => {
      return [];
    });
  }
  function setTransactionStatus(account, chainId, hash, status) {
    updateTransactions(account, chainId, (transactions) => {
      return transactions.map((transaction) => transaction.hash === hash ? { ...transaction, status } : transaction);
    });
  }
  async function waitForPendingTransactions(account, chainId) {
    await Promise.all(getTransactions(account, chainId).filter((transaction) => transaction.status === "pending").map(async (transaction) => {
      const { confirmations, hash } = transaction;
      const existingRequest = transactionRequestCache.get(hash);
      if (existingRequest) {
        return await existingRequest;
      }
      const requestPromise = provider.waitForTransaction(hash, confirmations).then(({ status }) => {
        transactionRequestCache.delete(hash);
        if (status === void 0) {
          return;
        }
        setTransactionStatus(account, chainId, hash, status === 0 ? "failed" : "confirmed");
      });
      transactionRequestCache.set(hash, requestPromise);
      return await requestPromise;
    }));
  }
  function updateTransactions(account, chainId, updateFn) {
    var _a, _b;
    data = loadData();
    data[account] = (_a = data[account]) != null ? _a : {};
    let completedTransactionCount = 0;
    const MAX_COMPLETED_TRANSACTIONS = 10;
    const transactions = updateFn((_b = data[account][chainId]) != null ? _b : []).filter(({ status }) => {
      return status === "pending" ? true : completedTransactionCount++ <= MAX_COMPLETED_TRANSACTIONS;
    });
    data[account][chainId] = transactions.length > 0 ? transactions : void 0;
    persistData();
    notifyListeners();
    waitForPendingTransactions(account, chainId);
  }
  function persistData() {
    localStorage.setItem(storageKey, JSON.stringify(data));
  }
  function notifyListeners() {
    listeners.forEach((listener) => listener());
  }
  function onChange(fn) {
    listeners.add(fn);
    return () => {
      listeners.delete(fn);
    };
  }
  return {
    addTransaction,
    clearTransactions,
    getTransactions,
    onChange,
    setProvider,
    waitForPendingTransactions
  };
}

// src/transactions/TransactionStoreContext.tsx
var storeSingleton;
var TransactionStoreContext = createContext2(null);
function TransactionStoreProvider({
  children
}) {
  const provider = useProvider();
  const address = useAddress();
  const chainId = useChainId();
  const [store] = useState(() => storeSingleton != null ? storeSingleton : storeSingleton = createTransactionStore({ provider }));
  useEffect3(() => {
    store.setProvider(provider);
  }, [store, provider]);
  useEffect3(() => {
    if (address && chainId) {
      store.waitForPendingTransactions(address, chainId);
    }
  }, [store, address, chainId]);
  return /* @__PURE__ */ React6.createElement(TransactionStoreContext.Provider, {
    value: store
  }, children);
}
function useTransactionStore() {
  const store = useContext2(TransactionStoreContext);
  if (!store) {
    throw new Error("Transaction hooks must be used within RainbowKitProvider");
  }
  return store;
}

// src/transactions/useRecentTransactions.ts
function useRecentTransactions() {
  const store = useTransactionStore();
  const address = useAddress();
  const chainId = useChainId();
  const [transactions, setTransactions] = useState2(() => store && address && chainId ? store.getTransactions(address, chainId) : []);
  useEffect4(() => {
    if (store && address && chainId) {
      setTransactions(store.getTransactions(address, chainId));
      return store.onChange(() => {
        setTransactions(store.getTransactions(address, chainId));
      });
    }
  }, [store, address, chainId]);
  return transactions;
}

// src/utils/isNotNullish.ts
function isNotNullish(value) {
  return value != null;
}

// src/wallets/useWalletConnectors.ts
import { useConnect } from "wagmi";

// src/utils/indexBy.ts
function indexBy(items, getKey) {
  const indexedItems = {};
  items.forEach((item) => {
    const key = getKey(item);
    if (!key) {
      return;
    }
    indexedItems[key] = item;
  });
  return indexedItems;
}

// src/wallets/recentWalletIds.ts
var storageKey2 = "rk-recent";
function safeParseJsonArray(string) {
  try {
    const value = string ? JSON.parse(string) : [];
    return Array.isArray(value) ? value : [];
  } catch (err) {
    return [];
  }
}
function getRecentWalletIds() {
  return typeof localStorage !== "undefined" ? safeParseJsonArray(localStorage.getItem(storageKey2)) : [];
}
function dedupe(array) {
  return [...new Set(array)];
}
function addRecentWalletId(walletId) {
  const newValue = dedupe([walletId, ...getRecentWalletIds()]);
  localStorage.setItem(storageKey2, JSON.stringify(newValue));
}

// src/wallets/useWalletConnectors.ts
function useWalletConnectors() {
  const { connectAsync, connectors: defaultConnectors } = useConnect();
  async function connectWallet(walletId, connector) {
    const result = await connectAsync(connector);
    if (result) {
      addRecentWalletId(walletId);
    }
    return result;
  }
  const connectorByWalletId = indexBy(defaultConnectors, (connector) => {
    var _a;
    return (_a = connector._wallet) == null ? void 0 : _a.id;
  });
  const MAX_RECENT_WALLETS = 3;
  const recentConnectors = getRecentWalletIds().map((walletId) => connectorByWalletId[walletId]).filter(isNotNullish).slice(0, MAX_RECENT_WALLETS);
  const connectors = [
    ...recentConnectors,
    ...defaultConnectors.filter((connector) => !recentConnectors.includes(connector))
  ];
  return connectors.map((connector) => {
    var _a;
    const wallet = connector._wallet;
    if (!wallet) {
      return null;
    }
    const recent = recentConnectors.includes(connector);
    return {
      ...wallet,
      connect: () => connectWallet(wallet.id, connector),
      groupName: recent ? "Recent" : wallet.groupName,
      onConnecting: (fn) => connector.on("message", ({ type }) => type === "connecting" ? fn() : void 0),
      ready: ((_a = wallet.installed) != null ? _a : true) && connector.ready,
      recent,
      showWalletConnectModal: wallet.walletConnectModalConnector ? async () => {
        try {
          await connectWallet(wallet.id, wallet.walletConnectModalConnector);
        } catch (err) {
          const isUserRejection = err.name === "UserRejectedRequestError";
          if (!isUserRejection) {
            throw err;
          }
        }
      } : void 0
    };
  }).filter(isNotNullish);
}

// src/components/AccountModal/AccountModal.tsx
import React24 from "react";

// src/components/Dialog/Dialog.tsx
import React9, {
  useCallback as useCallback2,
  useEffect as useEffect6
} from "react";
import { createPortal } from "react-dom";
import { RemoveScroll } from "react-remove-scroll";

// src/components/RainbowKitProvider/RainbowKitProvider.tsx
import React7, { createContext as createContext6, useContext as useContext3, useMemo as useMemo4 } from "react";

// src/css/cssObjectFromTheme.ts
import { assignInlineVars } from "@vanilla-extract/dynamic";
var resolveThemeVars = (theme) => typeof theme === "function" ? theme() : theme;
function cssObjectFromTheme(theme, { extends: baseTheme } = {}) {
  const resolvedThemeVars = {
    ...assignInlineVars(themeVars, resolveThemeVars(theme))
  };
  if (!baseTheme) {
    return resolvedThemeVars;
  }
  const resolvedBaseThemeVars = assignInlineVars(themeVars, resolveThemeVars(baseTheme));
  const filteredVars = Object.fromEntries(Object.entries(resolvedThemeVars).filter(([varName, value]) => value !== resolvedBaseThemeVars[varName]));
  return filteredVars;
}

// src/css/cssStringFromTheme.ts
function cssStringFromTheme(theme, options = {}) {
  return Object.entries(cssObjectFromTheme(theme, options)).map(([key, value]) => `${key}:${value};`).join("");
}

// src/components/RainbowKitProvider/AppContext.ts
import { createContext as createContext3 } from "react";
var defaultAppInfo = {
  appName: void 0,
  learnMoreUrl: "https://learn.rainbow.me/what-is-a-cryptoweb3-wallet-actually"
};
var AppContext = createContext3(defaultAppInfo);

// src/components/RainbowKitProvider/CoolModeContext.ts
import { createContext as createContext4 } from "react";
var CoolModeContext = createContext4(false);

// src/components/RainbowKitProvider/ShowRecentTransactionsContext.ts
import { createContext as createContext5 } from "react";
var ShowRecentTransactionsContext = createContext5(false);

// src/components/RainbowKitProvider/provideRainbowKitChains.ts
var arbitrumIcon = {
  iconBackground: "#96bedc",
  iconUrl: async () => (await import("./arbitrum-RJBKFWDE.js")).default
};
var avalancheIcon = {
  iconBackground: "#e84141",
  iconUrl: async () => (await import("./avalanche-GTODSNSK.js")).default
};
var ethereumIcon = {
  iconBackground: "#484c50",
  iconUrl: async () => (await import("./ethereum-TCZSTHSI.js")).default
};
var hardhatIcon = {
  iconBackground: "#f9f7ec",
  iconUrl: async () => (await import("./hardhat-VJNBZ2P3.js")).default
};
var optimismIcon = {
  iconBackground: "#ff5a57",
  iconUrl: async () => (await import("./optimism-3QUIF6AD.js")).default
};
var polygonIcon = {
  iconBackground: "#9f71ec",
  iconUrl: async () => (await import("./polygon-JVGMAJ3S.js")).default
};
var chainMetadataByName = {
  arbitrum: { chainId: 42161, ...arbitrumIcon },
  arbitrumRinkeby: { chainId: 421611, ...arbitrumIcon },
  avalanche: { chainId: 43114, ...avalancheIcon },
  avalancheFuji: { chainId: 43113, ...avalancheIcon },
  goerli: { chainId: 5, ...ethereumIcon },
  hardhat: { chainId: 31337, ...hardhatIcon },
  kovan: { chainId: 42, ...ethereumIcon },
  localhost: { chainId: 1337, ...ethereumIcon },
  mainnet: { chainId: 1, ...ethereumIcon },
  optimism: { chainId: 10, ...optimismIcon },
  optimismKovan: { chainId: 69, ...optimismIcon },
  polygon: { chainId: 137, ...polygonIcon },
  polygonMumbai: { chainId: 80001, ...polygonIcon },
  rinkeby: { chainId: 4, ...ethereumIcon },
  ropsten: { chainId: 3, ...ethereumIcon }
};
var chainMetadataById = Object.fromEntries(Object.values(chainMetadataByName).filter(isNotNullish).map(({ chainId, ...metadata }) => [chainId, metadata]));
var provideRainbowKitChains = (chains) => chains.map((chain) => {
  var _a;
  return {
    ...(_a = chainMetadataById[chain.id]) != null ? _a : {},
    ...chain
  };
});

// src/components/RainbowKitProvider/RainbowKitProvider.tsx
var ThemeIdContext = createContext6(void 0);
var anonymousDataAttribute = "data-rk";
var idDataAttribute = "data-rk-id";
var createThemeRootProps = (id) => id ? { [idDataAttribute]: id } : { [anonymousDataAttribute]: "" };
var createThemeRootSelector = (id) => id ? `[${idDataAttribute}="${id}"]` : `[${anonymousDataAttribute}]`;
var useThemeRootProps = () => {
  const id = useContext3(ThemeIdContext);
  return createThemeRootProps(id);
};
var defaultTheme = lightTheme();
function RainbowKitProvider({
  chains,
  id,
  theme = defaultTheme,
  children,
  appInfo,
  showRecentTransactions = false,
  coolMode = false
}) {
  const rainbowkitChains = useMemo4(() => provideRainbowKitChains(chains), [chains]);
  if (typeof theme === "function") {
    throw new Error('A theme function was provided to the "theme" prop instead of a theme object. You must execute this function to get the resulting theme object.');
  }
  const selector = createThemeRootSelector(id);
  const appContext = {
    ...defaultAppInfo,
    ...appInfo
  };
  return /* @__PURE__ */ React7.createElement(RainbowKitChainContext.Provider, {
    value: rainbowkitChains
  }, /* @__PURE__ */ React7.createElement(CoolModeContext.Provider, {
    value: coolMode
  }, /* @__PURE__ */ React7.createElement(ShowRecentTransactionsContext.Provider, {
    value: showRecentTransactions
  }, /* @__PURE__ */ React7.createElement(TransactionStoreProvider, null, /* @__PURE__ */ React7.createElement(AppContext.Provider, {
    value: appContext
  }, /* @__PURE__ */ React7.createElement(ThemeIdContext.Provider, {
    value: id
  }, theme ? /* @__PURE__ */ React7.createElement("div", {
    ...createThemeRootProps(id)
  }, /* @__PURE__ */ React7.createElement("style", null, [
    `${selector}{${cssStringFromTheme("lightMode" in theme ? theme.lightMode : theme)}}`,
    "darkMode" in theme ? `@media(prefers-color-scheme:dark){${selector}{${cssStringFromTheme(theme.darkMode, { extends: theme.lightMode })}}}` : null
  ].join("")), children) : children))))));
}

// src/components/Dialog/Dialog.css.ts
var content = "_9pm4ki5 ju367vp ju367v1d ju367v8j";
var overlay = "_9pm4ki3 ju367vaj ju367vp ju367v2v ju367v8i";

// src/components/Dialog/FocusTrap.tsx
import React8, { useCallback, useEffect as useEffect5, useRef } from "react";
var moveFocusWithin = (element2, position) => {
  const focusableElements = element2.querySelectorAll("button:not(:disabled), a[href]");
  if (focusableElements.length === 0)
    return;
  focusableElements[position === "end" ? focusableElements.length - 1 : 0].focus();
};
function FocusTrap(props) {
  const contentRef = useRef(null);
  useEffect5(() => {
    const previouslyActiveElement = document.activeElement;
    return () => {
      var _a;
      (_a = previouslyActiveElement.focus) == null ? void 0 : _a.call(previouslyActiveElement);
    };
  }, []);
  useEffect5(() => {
    if (contentRef.current) {
      const elementToFocus = contentRef.current.querySelector("[data-auto-focus]");
      if (elementToFocus) {
        elementToFocus.focus();
      } else {
        contentRef.current.focus();
      }
    }
  }, [contentRef]);
  return /* @__PURE__ */ React8.createElement(React8.Fragment, null, /* @__PURE__ */ React8.createElement("div", {
    onFocus: useCallback(() => contentRef.current && moveFocusWithin(contentRef.current, "end"), []),
    tabIndex: 0
  }), /* @__PURE__ */ React8.createElement("div", {
    ref: contentRef,
    style: { outline: "none" },
    tabIndex: -1,
    ...props
  }), /* @__PURE__ */ React8.createElement("div", {
    onFocus: useCallback(() => contentRef.current && moveFocusWithin(contentRef.current, "start"), []),
    tabIndex: 0
  }));
}

// src/components/Dialog/Dialog.tsx
var stopPropagation = (event) => event.stopPropagation();
function Dialog({ children, onClose, open, titleId }) {
  useEffect6(() => {
    const handleEscape = (event) => open && event.key === "Escape" && onClose();
    document.addEventListener("keydown", handleEscape);
    return () => document.removeEventListener("keydown", handleEscape);
  }, [open, onClose]);
  const handleBackdropClick = useCallback2(() => onClose(), [onClose]);
  const themeRootProps = useThemeRootProps();
  const mobile = isMobile();
  return /* @__PURE__ */ React9.createElement(React9.Fragment, null, open ? createPortal(/* @__PURE__ */ React9.createElement(RemoveScroll, null, /* @__PURE__ */ React9.createElement(Box, {
    ...themeRootProps,
    alignItems: mobile ? "flex-end" : "center",
    "aria-labelledby": titleId,
    "aria-modal": true,
    className: overlay,
    onClick: handleBackdropClick,
    position: "fixed",
    role: "dialog"
  }, /* @__PURE__ */ React9.createElement(FocusTrap, {
    className: content,
    onClick: stopPropagation,
    role: "document"
  }, children))), document.body) : null);
}

// src/components/Dialog/DialogContent.tsx
import React10 from "react";

// src/components/Dialog/DialogContent.css.ts
var bottomSheetOverrides = "_1ckjpok6";
var dialogContent = "_1ckjpok1 ju367vam ju367vd1 ju367v12 ju367v16 ju367v18 ju367vdv ju367vp ju367v1d ju367v68 ju367v8j";
var dialogContentMobile = "_1ckjpok5 ju367v13";
var dialogContentWideDesktop = "_1ckjpok3 _1ckjpok1 ju367vam ju367vd1 ju367v12 ju367v16 ju367v18 ju367vdv ju367vp ju367v1d ju367v68 ju367v8j";
var dialogContentWideMobile = "_1ckjpok2 _1ckjpok1 ju367vam ju367vd1 ju367v12 ju367v16 ju367v18 ju367vdv ju367vp ju367v1d ju367v68 ju367v8j";

// src/components/Dialog/DialogContent.tsx
function DialogContent({
  bottomSheetOnMobile = false,
  children,
  marginTop,
  padding = "16",
  wide = false
}) {
  const mobile = isMobile();
  return /* @__PURE__ */ React10.createElement(Box, {
    marginTop
  }, /* @__PURE__ */ React10.createElement(Box, {
    className: [
      wide ? mobile ? dialogContentWideMobile : dialogContentWideDesktop : dialogContent,
      mobile ? dialogContentMobile : null,
      mobile && bottomSheetOnMobile ? bottomSheetOverrides : null
    ].join(" ")
  }, /* @__PURE__ */ React10.createElement(Box, {
    padding
  }, children)));
}

// src/components/ProfileDetails/ProfileDetails.tsx
import React23, { useCallback as useCallback4, useContext as useContext5, useEffect as useEffect7, useState as useState3 } from "react";

// src/components/CloseButton/CloseButton.tsx
import React12 from "react";

// src/components/Icons/Close.tsx
import React11 from "react";
var CloseIcon = () => {
  return isMobile() ? /* @__PURE__ */ React11.createElement("svg", {
    fill: "none",
    height: "11.5",
    viewBox: "0 0 11.5 11.5",
    width: "11.5",
    xmlns: "http://www.w3.org/2000/svg"
  }, /* @__PURE__ */ React11.createElement("path", {
    d: "M2.13388 0.366117C1.64573 -0.122039 0.854272 -0.122039 0.366117 0.366117C-0.122039 0.854272 -0.122039 1.64573 0.366117 2.13388L3.98223 5.75L0.366117 9.36612C-0.122039 9.85427 -0.122039 10.6457 0.366117 11.1339C0.854272 11.622 1.64573 11.622 2.13388 11.1339L5.75 7.51777L9.36612 11.1339C9.85427 11.622 10.6457 11.622 11.1339 11.1339C11.622 10.6457 11.622 9.85427 11.1339 9.36612L7.51777 5.75L11.1339 2.13388C11.622 1.64573 11.622 0.854272 11.1339 0.366117C10.6457 -0.122039 9.85427 -0.122039 9.36612 0.366117L5.75 3.98223L2.13388 0.366117Z",
    fill: "currentColor"
  })) : /* @__PURE__ */ React11.createElement("svg", {
    fill: "none",
    height: "10",
    viewBox: "0 0 10 10",
    width: "10",
    xmlns: "http://www.w3.org/2000/svg"
  }, /* @__PURE__ */ React11.createElement("path", {
    d: "M1.70711 0.292893C1.31658 -0.0976311 0.683417 -0.0976311 0.292893 0.292893C-0.0976311 0.683417 -0.0976311 1.31658 0.292893 1.70711L3.58579 5L0.292893 8.29289C-0.0976311 8.68342 -0.0976311 9.31658 0.292893 9.70711C0.683417 10.0976 1.31658 10.0976 1.70711 9.70711L5 6.41421L8.29289 9.70711C8.68342 10.0976 9.31658 10.0976 9.70711 9.70711C10.0976 9.31658 10.0976 8.68342 9.70711 8.29289L6.41421 5L9.70711 1.70711C10.0976 1.31658 10.0976 0.683417 9.70711 0.292893C9.31658 -0.0976311 8.68342 -0.0976311 8.29289 0.292893L5 3.58579L1.70711 0.292893Z",
    fill: "currentColor"
  }));
};

// src/components/CloseButton/CloseButton.tsx
var CloseButton = ({
  onClose,
  style
}) => {
  const mobile = isMobile();
  return /* @__PURE__ */ React12.createElement(Box, {
    alignItems: "center",
    as: "button",
    background: "closeButtonBackground",
    borderColor: "actionButtonBorder",
    borderRadius: "full",
    borderStyle: "solid",
    borderWidth: mobile ? "0" : "1",
    color: "closeButton",
    display: "flex",
    height: mobile ? "30" : "28",
    justifyContent: "center",
    onClick: onClose,
    style: { willChange: "transform", ...style != null ? style : {} },
    transform: { active: "shrinkSm", hover: "growLg" },
    transition: "default",
    width: mobile ? "30" : "28"
  }, /* @__PURE__ */ React12.createElement(CloseIcon, null));
};

// src/components/ConnectButton/formatAddress.ts
function formatAddress(address) {
  const leadingChars = 4;
  const trailingChars = 4;
  return address.length < leadingChars + trailingChars ? address : `${address.substring(0, leadingChars)}\u2026${address.substring(address.length - trailingChars)}`;
}

// src/components/ConnectButton/formatENS.ts
function formatENS(name) {
  const parts = name.split(".");
  const last = parts.pop();
  if (parts.join(".").length > 24) {
    return `${parts.join(".").substring(0, 24)}...`;
  }
  return `${parts.join(".")}.${last}`;
}

// src/components/Icons/Copied.tsx
import React13 from "react";
var CopiedIcon = () => /* @__PURE__ */ React13.createElement("svg", {
  fill: "none",
  height: "13",
  viewBox: "0 0 13 13",
  width: "13",
  xmlns: "http://www.w3.org/2000/svg"
}, /* @__PURE__ */ React13.createElement("path", {
  d: "M4.94568 12.2646C5.41052 12.2646 5.77283 12.0869 6.01892 11.7109L12.39 1.96973C12.5677 1.69629 12.6429 1.44336 12.6429 1.2041C12.6429 0.561523 12.1644 0.0966797 11.5082 0.0966797C11.057 0.0966797 10.7767 0.260742 10.5033 0.691406L4.9115 9.50977L2.07458 5.98926C1.82166 5.68848 1.54822 5.55176 1.16541 5.55176C0.502319 5.55176 0.0238037 6.02344 0.0238037 6.66602C0.0238037 6.95312 0.112671 7.20605 0.358765 7.48633L3.88611 11.7588C4.18005 12.1074 4.50818 12.2646 4.94568 12.2646Z",
  fill: "currentColor"
}));

// src/components/Icons/Copy.tsx
import React14 from "react";
var CopyIcon = () => /* @__PURE__ */ React14.createElement("svg", {
  fill: "none",
  height: "16",
  viewBox: "0 0 17 16",
  width: "17",
  xmlns: "http://www.w3.org/2000/svg"
}, /* @__PURE__ */ React14.createElement("path", {
  d: "M3.04236 12.3027H4.18396V13.3008C4.18396 14.8525 5.03845 15.7002 6.59705 15.7002H13.6244C15.183 15.7002 16.0375 14.8525 16.0375 13.3008V6.24609C16.0375 4.69434 15.183 3.84668 13.6244 3.84668H12.4828V2.8418C12.4828 1.29688 11.6283 0.442383 10.0697 0.442383H3.04236C1.48376 0.442383 0.629272 1.29004 0.629272 2.8418V9.90332C0.629272 11.4551 1.48376 12.3027 3.04236 12.3027ZM3.23376 10.5391C2.68689 10.5391 2.39294 10.2656 2.39294 9.68457V3.06055C2.39294 2.47949 2.68689 2.21289 3.23376 2.21289H9.8783C10.4252 2.21289 10.7191 2.47949 10.7191 3.06055V3.84668H6.59705C5.03845 3.84668 4.18396 4.69434 4.18396 6.24609V10.5391H3.23376ZM6.78845 13.9365C6.24158 13.9365 5.94763 13.6699 5.94763 13.0889V6.45801C5.94763 5.87695 6.24158 5.61035 6.78845 5.61035H13.433C13.9799 5.61035 14.2738 5.87695 14.2738 6.45801V13.0889C14.2738 13.6699 13.9799 13.9365 13.433 13.9365H6.78845Z",
  fill: "currentColor"
}));

// src/components/Icons/Disconnect.tsx
import React15 from "react";
var DisconnectIcon = () => /* @__PURE__ */ React15.createElement("svg", {
  fill: "none",
  height: "16",
  viewBox: "0 0 18 16",
  width: "18",
  xmlns: "http://www.w3.org/2000/svg"
}, /* @__PURE__ */ React15.createElement("path", {
  d: "M2.67834 15.5908H9.99963C11.5514 15.5908 12.399 14.7432 12.399 13.1777V10.2656H10.6354V12.9863C10.6354 13.5332 10.3688 13.8271 9.78772 13.8271H2.89026C2.3092 13.8271 2.0426 13.5332 2.0426 12.9863V3.15625C2.0426 2.60254 2.3092 2.30859 2.89026 2.30859H9.78772C10.3688 2.30859 10.6354 2.60254 10.6354 3.15625V5.89746H12.399V2.95801C12.399 1.39941 11.5514 0.544922 9.99963 0.544922H2.67834C1.12659 0.544922 0.278931 1.39941 0.278931 2.95801V13.1777C0.278931 14.7432 1.12659 15.5908 2.67834 15.5908ZM7.43616 8.85059H14.0875L15.0924 8.78906L14.566 9.14453L13.6842 9.96484C13.5406 10.1016 13.4586 10.2861 13.4586 10.4844C13.4586 10.8398 13.7321 11.168 14.1217 11.168C14.3199 11.168 14.4635 11.0928 14.6002 10.9561L16.7809 8.68652C16.986 8.48145 17.0543 8.27637 17.0543 8.06445C17.0543 7.85254 16.986 7.64746 16.7809 7.43555L14.6002 5.17285C14.4635 5.03613 14.3199 4.9541 14.1217 4.9541C13.7321 4.9541 13.4586 5.27539 13.4586 5.6377C13.4586 5.83594 13.5406 6.02734 13.6842 6.15723L14.566 6.98438L15.0924 7.33984L14.0875 7.27148H7.43616C7.01917 7.27148 6.65686 7.62012 6.65686 8.06445C6.65686 8.50195 7.01917 8.85059 7.43616 8.85059Z",
  fill: "currentColor"
}));

// src/components/Text/Text.tsx
import React16 from "react";
var Text = React16.forwardRef(({
  as = "div",
  children,
  className,
  color,
  font = "body",
  id,
  size = "16",
  tabIndex,
  textAlign = "inherit",
  weight = "regular"
}, ref) => {
  return /* @__PURE__ */ React16.createElement(Box, {
    as,
    className,
    color,
    fontFamily: font,
    fontSize: size,
    fontWeight: weight,
    id,
    ref,
    tabIndex,
    textAlign
  }, children);
});
Text.displayName = "Text";

// src/components/Txs/TxList.tsx
import React21, { useContext as useContext4 } from "react";
import { useNetwork as useNetwork3 } from "wagmi";

// src/transactions/useClearRecentTransactions.ts
import { useCallback as useCallback3 } from "react";
function useClearRecentTransactions() {
  const store = useTransactionStore();
  const address = useAddress();
  const chainId = useChainId();
  return useCallback3(() => {
    if (!address || !chainId) {
      throw new Error("No address or chain ID found");
    }
    store.clearTransactions(address, chainId);
  }, [store, address, chainId]);
}

// src/utils/chainToExplorerUrl.ts
var chainToExplorerUrl = (chain) => {
  var _a, _b;
  return (_b = (_a = chain == null ? void 0 : chain.blockExplorers) == null ? void 0 : _a.default) == null ? void 0 : _b.url;
};

// src/components/Icons/ExternalLink.tsx
import React17 from "react";
var ExternalLinkIcon = () => /* @__PURE__ */ React17.createElement("svg", {
  fill: "none",
  height: "19",
  viewBox: "0 0 20 19",
  width: "20",
  xmlns: "http://www.w3.org/2000/svg"
}, /* @__PURE__ */ React17.createElement("path", {
  d: "M10 18.9443C15.0977 18.9443 19.2812 14.752 19.2812 9.6543C19.2812 4.56543 15.0889 0.373047 10 0.373047C4.90234 0.373047 0.71875 4.56543 0.71875 9.6543C0.71875 14.752 4.91113 18.9443 10 18.9443ZM10 16.6328C6.1416 16.6328 3.03906 13.5215 3.03906 9.6543C3.03906 5.7959 6.13281 2.68457 10 2.68457C13.8584 2.68457 16.9697 5.7959 16.9697 9.6543C16.9785 13.5215 13.8672 16.6328 10 16.6328ZM12.7158 12.1416C13.2432 12.1416 13.5684 11.7549 13.5684 11.1836V7.19336C13.5684 6.44629 13.1377 6.05957 12.417 6.05957H8.40918C7.8291 6.05957 7.45117 6.38477 7.45117 6.91211C7.45117 7.43945 7.8291 7.77344 8.40918 7.77344H9.69238L10.7207 7.63281L9.53418 8.67871L6.73047 11.4912C6.53711 11.6758 6.41406 11.9395 6.41406 12.2031C6.41406 12.7832 6.85352 13.1699 7.39844 13.1699C7.68848 13.1699 7.92578 13.0732 8.1543 12.8623L10.9316 10.0762L11.9775 8.89844L11.8545 9.98828V11.1836C11.8545 11.7725 12.1885 12.1416 12.7158 12.1416Z",
  fill: "currentColor"
}));

// src/components/Txs/TxItem.tsx
import React20 from "react";
import { useNetwork as useNetwork2 } from "wagmi";

// src/components/Icons/Cancel.tsx
import React18 from "react";
var CancelIcon = () => /* @__PURE__ */ React18.createElement("svg", {
  fill: "none",
  height: "19",
  viewBox: "0 0 20 19",
  width: "20",
  xmlns: "http://www.w3.org/2000/svg"
}, /* @__PURE__ */ React18.createElement("path", {
  d: "M10 18.9443C15.0977 18.9443 19.2812 14.752 19.2812 9.6543C19.2812 4.56543 15.0889 0.373047 10 0.373047C4.90234 0.373047 0.71875 4.56543 0.71875 9.6543C0.71875 14.752 4.91113 18.9443 10 18.9443ZM10 16.6328C6.1416 16.6328 3.03906 13.5215 3.03906 9.6543C3.03906 5.7959 6.13281 2.68457 10 2.68457C13.8584 2.68457 16.9697 5.7959 16.9697 9.6543C16.9785 13.5215 13.8672 16.6328 10 16.6328ZM7.29297 13.3018C7.58301 13.3018 7.81152 13.2139 7.99609 13.0205L10 11.0166L12.0127 13.0205C12.1973 13.2051 12.4258 13.3018 12.707 13.3018C13.2432 13.3018 13.6562 12.8887 13.6562 12.3525C13.6562 12.0977 13.5508 11.8691 13.3662 11.6934L11.3535 9.67188L13.375 7.6416C13.5596 7.44824 13.6562 7.22852 13.6562 6.98242C13.6562 6.44629 13.2432 6.0332 12.7158 6.0332C12.4346 6.0332 12.2148 6.12109 12.0215 6.31445L10 8.32715L7.9873 6.32324C7.80273 6.12988 7.58301 6.04199 7.29297 6.04199C6.76562 6.04199 6.35254 6.45508 6.35254 6.99121C6.35254 7.2373 6.44922 7.46582 6.63379 7.6416L8.65527 9.67188L6.63379 11.6934C6.44922 11.8691 6.35254 12.1064 6.35254 12.3525C6.35254 12.8887 6.76562 13.3018 7.29297 13.3018Z",
  fill: "currentColor"
}));

// src/components/Icons/Success.tsx
import React19 from "react";
var SuccessIcon = () => /* @__PURE__ */ React19.createElement("svg", {
  fill: "none",
  height: "20",
  viewBox: "0 0 20 20",
  width: "20",
  xmlns: "http://www.w3.org/2000/svg"
}, /* @__PURE__ */ React19.createElement("path", {
  d: "M10 19.4443C15.0977 19.4443 19.2812 15.252 19.2812 10.1543C19.2812 5.06543 15.0889 0.873047 10 0.873047C4.90234 0.873047 0.71875 5.06543 0.71875 10.1543C0.71875 15.252 4.91113 19.4443 10 19.4443ZM10 17.1328C6.1416 17.1328 3.03906 14.0215 3.03906 10.1543C3.03906 6.2959 6.13281 3.18457 10 3.18457C13.8584 3.18457 16.9697 6.2959 16.9697 10.1543C16.9785 14.0215 13.8672 17.1328 10 17.1328ZM9.07715 14.3379C9.4375 14.3379 9.7627 14.1533 9.97363 13.8369L13.7441 8.00977C13.8848 7.79883 13.9814 7.5791 13.9814 7.36816C13.9814 6.84961 13.5244 6.48926 13.0322 6.48926C12.707 6.48926 12.4258 6.66504 12.2148 7.0166L9.05957 12.0967L7.5918 10.2949C7.37207 10.0225 7.13477 9.9082 6.84473 9.9082C6.33496 9.9082 5.92188 10.3125 5.92188 10.8223C5.92188 11.0684 6.00098 11.2793 6.18555 11.5078L8.1543 13.8545C8.40918 14.1709 8.70801 14.3379 9.07715 14.3379Z",
  fill: "currentColor"
}));

// src/components/Txs/TxItem.tsx
var getTxStatusIcon = (status) => {
  switch (status) {
    case "pending":
      return SpinnerIcon;
    case "confirmed":
      return SuccessIcon;
    case "failed":
      return CancelIcon;
    default:
      return SpinnerIcon;
  }
};
function TxItem({ tx }) {
  const mobile = isMobile();
  const Icon = getTxStatusIcon(tx.status);
  const color = tx.status === "failed" ? "error" : "accentColor";
  const { activeChain } = useNetwork2();
  const confirmationStatus = tx.status === "confirmed" ? "Confirmed" : tx.status === "failed" ? "Failed" : "Pending";
  const explorerLink = chainToExplorerUrl(activeChain);
  return /* @__PURE__ */ React20.createElement(React20.Fragment, null, /* @__PURE__ */ React20.createElement(Box, {
    ...explorerLink ? {
      as: "a",
      background: { hover: "profileForeground" },
      borderRadius: "menuButton",
      href: `${explorerLink}/tx/${tx.hash}`,
      rel: "noreferrer",
      target: "_blank",
      transform: { active: "shrink", hover: "grow" },
      transition: "default"
    } : {},
    color: "modalText",
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between",
    padding: "8"
  }, /* @__PURE__ */ React20.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "row",
    gap: mobile ? "16" : "14"
  }, /* @__PURE__ */ React20.createElement(Box, {
    color
  }, /* @__PURE__ */ React20.createElement(Icon, null)), /* @__PURE__ */ React20.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    gap: mobile ? "4" : "2"
  }, /* @__PURE__ */ React20.createElement(Box, null, /* @__PURE__ */ React20.createElement(Text, {
    color: "modalText",
    font: "body",
    size: mobile ? "16" : "14",
    weight: "bold"
  }, tx == null ? void 0 : tx.description)), /* @__PURE__ */ React20.createElement(Box, null, /* @__PURE__ */ React20.createElement(Text, {
    color: tx.status === "pending" ? "modalTextSecondary" : color,
    font: "body",
    size: "14",
    weight: mobile ? "medium" : "regular"
  }, confirmationStatus)))), explorerLink && /* @__PURE__ */ React20.createElement(Box, {
    alignItems: "center",
    color: "modalTextDim",
    display: "flex"
  }, /* @__PURE__ */ React20.createElement(ExternalLinkIcon, null))), mobile && /* @__PURE__ */ React20.createElement(Box, {
    background: "generalBorderDim",
    height: "1",
    marginLeft: "44",
    marginRight: "8",
    marginY: "2"
  }));
}

// src/components/Txs/TxList.tsx
var NUMBER_OF_VISIBLE_TXS = 3;
function TxList({ accountData }) {
  const recentTransactions = useRecentTransactions();
  const clearRecentTransactions = useClearRecentTransactions();
  const { activeChain } = useNetwork3();
  const address = accountData == null ? void 0 : accountData.address;
  const explorerLink = chainToExplorerUrl(activeChain);
  const visibleTxs = recentTransactions.slice(0, NUMBER_OF_VISIBLE_TXS);
  const hasTransactions = visibleTxs.length > 0;
  const mobile = isMobile();
  const { appName } = useContext4(AppContext);
  return /* @__PURE__ */ React21.createElement(React21.Fragment, null, /* @__PURE__ */ React21.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    gap: "10",
    paddingBottom: "4",
    paddingTop: "16",
    paddingX: mobile ? "8" : "18"
  }, hasTransactions && /* @__PURE__ */ React21.createElement(Box, {
    paddingBottom: mobile ? "4" : "0",
    paddingTop: mobile ? "0" : "8",
    paddingX: mobile ? "12" : "6"
  }, /* @__PURE__ */ React21.createElement(Box, {
    display: "flex",
    gap: "4",
    justifyContent: "space-between"
  }, /* @__PURE__ */ React21.createElement(Text, {
    color: "modalTextSecondary",
    size: mobile ? "16" : "14",
    weight: "semibold"
  }, "Recent Transactions"), /* @__PURE__ */ React21.createElement(Box, {
    as: "button",
    background: {
      hover: "profileForeground"
    },
    borderRadius: "actionButton",
    onClick: clearRecentTransactions,
    padding: "6",
    paddingX: "10",
    style: {
      marginBottom: -6,
      marginLeft: -10,
      marginRight: -10,
      marginTop: -6
    },
    transform: {
      active: "shrink",
      hover: "grow"
    },
    transition: "default",
    type: "button"
  }, /* @__PURE__ */ React21.createElement(Text, {
    color: "modalTextSecondary",
    size: mobile ? "16" : "14",
    weight: "semibold"
  }, "Clear All")))), /* @__PURE__ */ React21.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    gap: "4"
  }, hasTransactions ? visibleTxs.map((tx) => /* @__PURE__ */ React21.createElement(TxItem, {
    key: tx.hash,
    tx
  })) : /* @__PURE__ */ React21.createElement(React21.Fragment, null, /* @__PURE__ */ React21.createElement(Box, {
    padding: mobile ? "12" : "8"
  }, /* @__PURE__ */ React21.createElement(Text, {
    color: "modalTextDim",
    size: mobile ? "16" : "14",
    weight: mobile ? "medium" : "bold"
  }, appName != null ? appName : "Your", " transactions will appear here...")), mobile && /* @__PURE__ */ React21.createElement(Box, {
    background: "generalBorderDim",
    height: "1",
    marginX: "12",
    marginY: "8"
  })))), explorerLink && /* @__PURE__ */ React21.createElement(Box, {
    paddingBottom: "18",
    paddingX: mobile ? "8" : "18"
  }, /* @__PURE__ */ React21.createElement(Box, {
    alignItems: "center",
    as: "a",
    background: { hover: "profileForeground" },
    borderRadius: "menuButton",
    color: "modalTextDim",
    display: "flex",
    flexDirection: "row",
    href: `${explorerLink}/address/${address}`,
    justifyContent: "space-between",
    paddingX: "8",
    paddingY: "12",
    rel: "noreferrer",
    style: { willChange: "transform" },
    target: "_blank",
    transform: {
      active: "shrink",
      hover: "grow"
    },
    transition: "default",
    ...mobile ? { paddingLeft: "12" } : {}
  }, /* @__PURE__ */ React21.createElement(Text, {
    color: "modalText",
    font: "body",
    size: mobile ? "16" : "14",
    weight: mobile ? "semibold" : "bold"
  }, "View more on Explorer"), /* @__PURE__ */ React21.createElement(ExternalLinkIcon, null))));
}

// src/components/ProfileDetails/ProfileDetailsAction.tsx
import React22 from "react";
function ProfileDetailsAction({
  action,
  icon,
  label,
  url
}) {
  const urlProps = url ? {
    href: url,
    rel: "noreferrer",
    target: "_blank"
  } : {};
  const mobile = isMobile();
  return /* @__PURE__ */ React22.createElement(Box, {
    as: url ? "a" : "button",
    onClick: action,
    ...urlProps,
    background: {
      base: "profileAction",
      ...!mobile ? { hover: "profileActionHover" } : {}
    },
    borderRadius: "menuButton",
    boxShadow: "profileDetailsAction",
    padding: mobile ? "8" : "10",
    style: { flexBasis: 0, flexGrow: 1, willChange: "transform" },
    transform: {
      active: "shrinkSm",
      ...!mobile ? { hover: "grow" } : {}
    },
    transition: "default"
  }, /* @__PURE__ */ React22.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    gap: "2",
    justifyContent: "center"
  }, /* @__PURE__ */ React22.createElement(Box, {
    color: "modalText",
    height: "max"
  }, icon), /* @__PURE__ */ React22.createElement(Box, null, /* @__PURE__ */ React22.createElement(Text, {
    color: "modalText",
    size: mobile ? "12" : "13",
    weight: "semibold"
  }, label))));
}

// src/components/ProfileDetails/ProfileDetails.tsx
function ProfileDetails({
  accountData,
  balanceData,
  ensAvatar,
  ensName,
  onClose,
  onDisconnect
}) {
  const showRecentTransactions = useContext5(ShowRecentTransactionsContext);
  const [copiedAddress, setCopiedAddress] = useState3(false);
  const copyAddressAction = useCallback4(() => {
    if (accountData == null ? void 0 : accountData.address) {
      navigator.clipboard.writeText(accountData == null ? void 0 : accountData.address);
      setCopiedAddress(true);
    }
  }, [accountData == null ? void 0 : accountData.address]);
  useEffect7(() => {
    if (copiedAddress) {
      const timer = setTimeout(() => {
        setCopiedAddress(false);
      }, 1500);
      return () => clearTimeout(timer);
    }
  }, [copiedAddress]);
  if (!(accountData == null ? void 0 : accountData.address)) {
    return null;
  }
  const accountName = ensName ? formatENS(ensName) : formatAddress(accountData.address);
  const ethBalance = balanceData == null ? void 0 : balanceData.formatted;
  const balance = Number(ethBalance).toPrecision(3);
  const titleId = "rk_profile_title";
  const mobile = isMobile();
  return /* @__PURE__ */ React23.createElement(React23.Fragment, null, /* @__PURE__ */ React23.createElement(Box, {
    display: "flex",
    flexDirection: "column"
  }, /* @__PURE__ */ React23.createElement(Box, {
    background: "profileForeground",
    padding: "16"
  }, /* @__PURE__ */ React23.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    gap: mobile ? "16" : "12",
    justifyContent: "center",
    margin: "8",
    style: { textAlign: "center" }
  }, /* @__PURE__ */ React23.createElement(CloseButton, {
    onClose,
    style: {
      position: "absolute",
      right: 16,
      top: 16,
      willChange: "transform"
    }
  }), " ", /* @__PURE__ */ React23.createElement(Box, {
    marginTop: mobile ? "24" : "0"
  }, /* @__PURE__ */ React23.createElement(Avatar, {
    address: accountData.address,
    imageUrl: ensAvatar,
    size: mobile ? 82 : 74
  })), /* @__PURE__ */ React23.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    gap: mobile ? "4" : "0",
    textAlign: "center"
  }, /* @__PURE__ */ React23.createElement(Box, {
    textAlign: "center"
  }, /* @__PURE__ */ React23.createElement(Text, {
    as: "h1",
    color: "modalText",
    id: titleId,
    size: mobile ? "20" : "18",
    weight: "heavy"
  }, accountName)), balanceData && /* @__PURE__ */ React23.createElement(Box, {
    textAlign: "center"
  }, /* @__PURE__ */ React23.createElement(Text, {
    as: "h1",
    color: "modalTextSecondary",
    id: titleId,
    size: mobile ? "16" : "14",
    weight: "semibold"
  }, balance, " ", balanceData.symbol)))), /* @__PURE__ */ React23.createElement(Box, {
    display: "flex",
    flexDirection: "row",
    gap: "8",
    margin: "2",
    marginTop: "16"
  }, /* @__PURE__ */ React23.createElement(ProfileDetailsAction, {
    action: copyAddressAction,
    icon: copiedAddress ? /* @__PURE__ */ React23.createElement(CopiedIcon, null) : /* @__PURE__ */ React23.createElement(CopyIcon, null),
    label: copiedAddress ? "Copied!" : "Copy Address"
  }), /* @__PURE__ */ React23.createElement(ProfileDetailsAction, {
    action: onDisconnect,
    icon: /* @__PURE__ */ React23.createElement(DisconnectIcon, null),
    label: "Disconnect"
  }))), showRecentTransactions && /* @__PURE__ */ React23.createElement(React23.Fragment, null, /* @__PURE__ */ React23.createElement(Box, {
    background: "generalBorder",
    height: "1",
    marginTop: "-1"
  }), /* @__PURE__ */ React23.createElement(Box, null, /* @__PURE__ */ React23.createElement(TxList, {
    accountData
  })))));
}

// src/components/AccountModal/AccountModal.tsx
function AccountModal({
  accountData,
  balanceData,
  ensAvatar,
  ensName,
  onClose,
  onDisconnect,
  open
}) {
  if (!accountData) {
    return null;
  }
  const titleId = "rk_account_modal_title";
  return /* @__PURE__ */ React24.createElement(React24.Fragment, null, accountData && /* @__PURE__ */ React24.createElement(Dialog, {
    onClose,
    open,
    titleId
  }, /* @__PURE__ */ React24.createElement(DialogContent, {
    bottomSheetOnMobile: true,
    padding: "0"
  }, /* @__PURE__ */ React24.createElement(ProfileDetails, {
    accountData,
    balanceData,
    ensAvatar,
    ensName,
    onClose,
    onDisconnect
  }))));
}

// src/components/ChainModal/ChainModal.tsx
import React26, {
  Fragment,
  useCallback as useCallback5,
  useContext as useContext6,
  useEffect as useEffect8,
  useState as useState4
} from "react";
import { useConnect as useConnect2 } from "wagmi";

// src/components/MenuButton/MenuButton.tsx
import React25 from "react";

// src/components/MenuButton/MenuButton.css.ts
var MenuButtonClassName = "ju367v82 ju367v6e ju367v6y ju367v7i ju367vah ju367v11 ju367vfy ju367v8 ju367vc";
var MobileMenuButtonClassName = "v9horb3 ju367v83 ju367v6f ju367v6z ju367v7j";
var SelectedMenuButtonClassName = "ju367v82 ju367v6e ju367v6y ju367v7i ju367v94 ju367vdm ju367v11 ju367v16 ju367v18 ju367ve1 ju367ved";

// src/components/MenuButton/MenuButton.tsx
var MenuButton = React25.forwardRef(({
  as = "button",
  children,
  currentlySelected = false,
  onClick,
  ...urlProps
}, ref) => {
  const mobile = isMobile();
  return /* @__PURE__ */ React25.createElement(Box, {
    as,
    className: [
      currentlySelected ? SelectedMenuButtonClassName : MenuButtonClassName,
      mobile ? MobileMenuButtonClassName : null
    ],
    disabled: currentlySelected,
    onClick,
    ref,
    ...urlProps
  }, children);
});
MenuButton.displayName = "MenuButton";

// src/components/ChainModal/ChainModal.tsx
function ChainModal({
  activeChain,
  chains,
  networkError,
  onClose,
  onSwitchNetwork,
  open
}) {
  const { activeConnector } = useConnect2();
  const [switchingToChain, setSwitchingToChain] = useState4();
  const titleId = "rk_chain_modal_title";
  const mobile = isMobile();
  const rainbowkitChainsById = useRainbowKitChainsById();
  const stopSwitching = useCallback5(() => {
    setSwitchingToChain(null);
    onClose();
  }, [onClose]);
  useEffect8(() => {
    var _a;
    if (!activeConnector) {
      return;
    }
    const stopSwitching2 = () => {
      setSwitchingToChain(null);
      onClose();
    };
    let provider;
    (_a = activeConnector == null ? void 0 : activeConnector.getProvider) == null ? void 0 : _a.call(activeConnector).then((provider_) => {
      provider = provider_;
      provider.on("chainChanged", stopSwitching2);
    });
    return () => {
      provider == null ? void 0 : provider.removeListener("chainChanged", stopSwitching2);
    };
  }, [activeConnector, onClose, stopSwitching]);
  useEffect8(() => {
    if (networkError && networkError.name === "UserRejectedRequestError") {
      stopSwitching();
    }
  }, [networkError, stopSwitching]);
  const { appName } = useContext6(AppContext);
  if (!activeChain || !(activeChain == null ? void 0 : activeChain.id)) {
    return null;
  }
  return /* @__PURE__ */ React26.createElement(Dialog, {
    onClose,
    open,
    titleId
  }, /* @__PURE__ */ React26.createElement(DialogContent, {
    bottomSheetOnMobile: true
  }, /* @__PURE__ */ React26.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    gap: "14"
  }, /* @__PURE__ */ React26.createElement(Box, {
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between"
  }, mobile && /* @__PURE__ */ React26.createElement("div", null), /* @__PURE__ */ React26.createElement(Box, {
    paddingBottom: "0",
    paddingLeft: "10",
    paddingTop: "4"
  }, /* @__PURE__ */ React26.createElement(Text, {
    as: "h1",
    color: "modalText",
    id: titleId,
    size: mobile ? "20" : "18",
    weight: "heavy"
  }, "Switch Networks")), /* @__PURE__ */ React26.createElement(CloseButton, {
    onClose
  })), /* @__PURE__ */ React26.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    gap: "4",
    padding: "2"
  }, onSwitchNetwork ? chains.map((chain, idx) => {
    const isCurrentChain = chain.id === (activeChain == null ? void 0 : activeChain.id);
    const switching = chain.id === switchingToChain;
    const rainbowKitChain = rainbowkitChainsById[chain.id];
    const chainIconSize = mobile ? "36" : "28";
    const chainIconUrl = rainbowKitChain == null ? void 0 : rainbowKitChain.iconUrl;
    const chainIconBackground = rainbowKitChain == null ? void 0 : rainbowKitChain.iconBackground;
    return /* @__PURE__ */ React26.createElement(Fragment, {
      key: chain.id
    }, /* @__PURE__ */ React26.createElement(MenuButton, {
      currentlySelected: isCurrentChain,
      onClick: isCurrentChain ? void 0 : () => {
        setSwitchingToChain(chain.id);
        onSwitchNetwork(chain.id);
      }
    }, /* @__PURE__ */ React26.createElement(Box, {
      fontFamily: "body",
      fontSize: "16",
      fontWeight: "bold"
    }, /* @__PURE__ */ React26.createElement(Box, {
      alignItems: "center",
      display: "flex",
      flexDirection: "row",
      justifyContent: "space-between"
    }, /* @__PURE__ */ React26.createElement(Box, {
      alignItems: "center",
      display: "flex",
      flexDirection: "row",
      gap: "4",
      height: chainIconSize
    }, chainIconUrl ? /* @__PURE__ */ React26.createElement(Box, {
      height: "full",
      marginRight: "8"
    }, /* @__PURE__ */ React26.createElement(AsyncImage, {
      alt: chain.name,
      background: chainIconBackground,
      borderRadius: "full",
      height: chainIconSize,
      src: chainIconUrl,
      width: chainIconSize
    })) : null, /* @__PURE__ */ React26.createElement("div", null, chain.name)), isCurrentChain && /* @__PURE__ */ React26.createElement(Box, {
      alignItems: "center",
      display: "flex",
      flexDirection: "row",
      marginRight: "6"
    }, /* @__PURE__ */ React26.createElement(Text, {
      color: "accentColorForeground",
      size: "14",
      weight: "medium"
    }, "Connected"), /* @__PURE__ */ React26.createElement(Box, {
      background: "connectionIndicator",
      borderColor: "selectedOptionBorder",
      borderRadius: "full",
      borderStyle: "solid",
      borderWidth: "1",
      height: "8",
      marginLeft: "8",
      width: "8"
    })), switching && /* @__PURE__ */ React26.createElement(Box, {
      alignItems: "center",
      display: "flex",
      flexDirection: "row",
      marginRight: "6"
    }, /* @__PURE__ */ React26.createElement(Text, {
      color: "modalText",
      size: "14",
      weight: "medium"
    }, "Confirm in Wallet"), /* @__PURE__ */ React26.createElement(Box, {
      background: "standby",
      borderRadius: "full",
      height: "8",
      marginLeft: "8",
      width: "8"
    }))))), mobile && idx < (chains == null ? void 0 : chains.length) - 1 && /* @__PURE__ */ React26.createElement(Box, {
      background: "generalBorderDim",
      height: "1",
      marginX: "8"
    }));
  }) : /* @__PURE__ */ React26.createElement(Box, {
    background: "generalBorder",
    borderRadius: "menuButton",
    paddingX: "18",
    paddingY: "12"
  }, /* @__PURE__ */ React26.createElement(Text, {
    color: "modalText",
    size: "14",
    weight: "medium"
  }, "Your wallet does not support switching networks from", " ", appName != null ? appName : "this app", ". Try switching networks from within your wallet instead."))))));
}

// src/components/ConnectModal/ConnectModal.tsx
import React40 from "react";
import { useAccount as useAccount3 } from "wagmi";

// src/components/ConnectOptions/ConnectOptions.tsx
import React39 from "react";

// src/components/ConnectOptions/DesktopOptions.tsx
import React37, { Fragment as Fragment2, useEffect as useEffect12, useState as useState7 } from "react";

// src/utils/browsers.ts
import { detect as detect2 } from "detect-browser";
function isSafari() {
  const browser = detect2();
  return !!browser && browser.name === "safari";
}

// src/utils/groupBy.ts
function groupBy(items, getKey) {
  const groupedItems = {};
  items.forEach((item) => {
    const key = getKey(item);
    if (!key) {
      return;
    }
    if (!groupedItems[key]) {
      groupedItems[key] = [];
    }
    groupedItems[key].push(item);
  });
  return groupedItems;
}

// src/components/ConnectModal/ConnectModalIntro.tsx
import React30, { useContext as useContext7 } from "react";

// src/components/Button/ActionButton.tsx
import React27 from "react";
var sizeVariants = {
  large: {
    fontSize: "16",
    paddingX: "24",
    paddingY: "10"
  },
  medium: {
    fontSize: "14",
    height: "28",
    paddingX: "12",
    paddingY: "4"
  },
  small: {
    fontSize: "14",
    paddingX: "10",
    paddingY: "5"
  }
};
function ActionButton({
  href,
  label,
  onClick,
  size = "medium",
  target = "_blank",
  type = "primary"
}) {
  const isPrimary = type === "primary";
  const isNotLarge = size !== "large";
  const mobile = isMobile();
  const background = isPrimary ? "accentColor" : isNotLarge ? "actionButtonSecondaryBackground" : null;
  const { fontSize, height, paddingX, paddingY } = sizeVariants[size];
  const hasBorder = !mobile || !isNotLarge;
  return /* @__PURE__ */ React27.createElement(Box, {
    ...href ? { as: "a", href, rel: "noreferrer noopener", target } : { as: "button", type: "button" },
    ...hasBorder ? {
      borderColor: mobile && !isNotLarge && !isPrimary ? "actionButtonBorderMobile" : "actionButtonBorder",
      borderStyle: "solid",
      borderWidth: "1"
    } : {},
    borderRadius: "actionButton",
    display: "block",
    onClick,
    paddingX,
    paddingY,
    style: { willChange: "transform" },
    textAlign: "center",
    transform: { active: "shrinkSm", hover: "grow" },
    transition: "default",
    ...background ? { background } : {},
    ...height ? { height } : {}
  }, /* @__PURE__ */ React27.createElement(Text, {
    color: isPrimary ? "accentColorForeground" : "accentColor",
    size: fontSize,
    weight: "bold"
  }, label));
}

// src/components/Icons/Assets.tsx
import React28 from "react";
var src = async () => (await import("./assets-7THATBKH.js")).default;
var preloadAssetsIcon = () => loadImages(src);
var AssetsIcon = () => /* @__PURE__ */ React28.createElement(AsyncImage, {
  background: "#d0d5de",
  borderRadius: "10",
  height: "48",
  src,
  width: "48"
});

// src/components/Icons/Login.tsx
import React29 from "react";
var src2 = async () => (await import("./login-EJZTP74Q.js")).default;
var preloadLoginIcon = () => loadImages(src2);
var LoginIcon = () => /* @__PURE__ */ React29.createElement(AsyncImage, {
  background: "#d0d5de",
  borderRadius: "10",
  height: "48",
  src: src2,
  width: "48"
});

// src/components/ConnectModal/ConnectModalIntro.tsx
function ConnectModalIntro({ getWallet }) {
  const { learnMoreUrl } = useContext7(AppContext);
  return /* @__PURE__ */ React30.createElement(React30.Fragment, null, /* @__PURE__ */ React30.createElement(Box, {
    alignItems: "center",
    color: "accentColor",
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    style: { gap: 62 }
  }, /* @__PURE__ */ React30.createElement(Text, {
    color: "modalText",
    size: "18",
    weight: "heavy"
  }, "What is a Wallet?"), /* @__PURE__ */ React30.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    gap: "32",
    height: "full",
    justifyContent: "center",
    style: { maxWidth: 312 }
  }, /* @__PURE__ */ React30.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "row",
    gap: "16"
  }, /* @__PURE__ */ React30.createElement(Box, {
    borderRadius: "6",
    height: "48",
    minWidth: "48",
    width: "48"
  }, /* @__PURE__ */ React30.createElement(AssetsIcon, null)), /* @__PURE__ */ React30.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    gap: "4"
  }, /* @__PURE__ */ React30.createElement(Text, {
    color: "modalText",
    size: "14",
    weight: "bold"
  }, "A Home for your Digital Assets"), /* @__PURE__ */ React30.createElement(Text, {
    color: "modalTextSecondary",
    size: "14",
    weight: "medium"
  }, "Wallets are used to send, receive, store, and display digital assets like Ethereum and NFTs."))), /* @__PURE__ */ React30.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "row",
    gap: "16"
  }, /* @__PURE__ */ React30.createElement(Box, {
    borderRadius: "6",
    height: "48",
    minWidth: "48",
    width: "48"
  }, /* @__PURE__ */ React30.createElement(LoginIcon, null)), /* @__PURE__ */ React30.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    gap: "4"
  }, /* @__PURE__ */ React30.createElement(Text, {
    color: "modalText",
    size: "14",
    weight: "bold"
  }, "A New Way to Log In"), /* @__PURE__ */ React30.createElement(Text, {
    color: "modalTextSecondary",
    size: "14",
    weight: "medium"
  }, "Instead of creating new accounts and passwords on every website, just connect your wallet.")))), /* @__PURE__ */ React30.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    gap: "12",
    justifyContent: "center",
    marginTop: "8"
  }, /* @__PURE__ */ React30.createElement(ActionButton, {
    label: "Get a Wallet",
    onClick: getWallet
  }), /* @__PURE__ */ React30.createElement(Box, {
    as: "a",
    href: learnMoreUrl,
    paddingX: "12",
    paddingY: "4",
    rel: "noreferrer",
    style: { willChange: "transform" },
    target: "_blank",
    transform: { active: "shrink", hover: "grow" },
    transition: "default"
  }, /* @__PURE__ */ React30.createElement(Text, {
    color: "accentColor",
    size: "14",
    weight: "bold"
  }, "Learn More")))));
}

// src/components/Icons/Back.tsx
import React31 from "react";
var BackIcon = () => /* @__PURE__ */ React31.createElement("svg", {
  fill: "none",
  height: "17",
  viewBox: "0 0 11 17",
  width: "11",
  xmlns: "http://www.w3.org/2000/svg"
}, /* @__PURE__ */ React31.createElement("path", {
  d: "M0.99707 8.6543C0.99707 9.08496 1.15527 9.44531 1.51562 9.79688L8.16016 16.3096C8.43262 16.5732 8.74902 16.7051 9.13574 16.7051C9.90918 16.7051 10.5508 16.0811 10.5508 15.3076C10.5508 14.9121 10.3838 14.5605 10.0938 14.2705L4.30176 8.64551L10.0938 3.0293C10.3838 2.74805 10.5508 2.3877 10.5508 2.00098C10.5508 1.23633 9.90918 0.603516 9.13574 0.603516C8.74902 0.603516 8.43262 0.735352 8.16016 0.999023L1.51562 7.51172C1.15527 7.85449 1.00586 8.21484 0.99707 8.6543Z",
  fill: "currentColor"
}));

// src/components/ModalSelection/ModalSelection.tsx
import React32, { useState as useState5 } from "react";

// src/components/RainbowKitProvider/useCoolMode.ts
import { useContext as useContext8, useEffect as useEffect9, useRef as useRef2 } from "react";
var useCoolMode = (imageUrl) => {
  const ref = useRef2(null);
  const coolModeEnabled = useContext8(CoolModeContext);
  const resolvedImageUrl = useAsyncImage(imageUrl);
  useEffect9(() => {
    if (coolModeEnabled && ref.current && resolvedImageUrl) {
      return makeElementCool(ref.current, resolvedImageUrl);
    }
  }, [coolModeEnabled, resolvedImageUrl]);
  return ref;
};
var getContainer = () => {
  const id = "_rk_coolMode";
  const existingContainer = document.getElementById(id);
  if (existingContainer) {
    return existingContainer;
  }
  const container = document.createElement("div");
  container.setAttribute("id", id);
  container.setAttribute("style", [
    "overflow:hidden",
    "position:fixed",
    "height:100%",
    "top:0",
    "left:0",
    "right:0",
    "bottom:0",
    "pointer-events:none",
    "z-index:2147483647"
  ].join(";"));
  document.body.appendChild(container);
  return container;
};
var instanceCounter = 0;
function makeElementCool(element2, imageUrl) {
  instanceCounter++;
  const sizes = [15, 20, 25, 35, 45];
  const limit = 35;
  let particles = [];
  let autoAddParticle = false;
  let mouseX = 0;
  let mouseY = 0;
  const container = getContainer();
  function createParticle() {
    const size = sizes[Math.floor(Math.random() * sizes.length)];
    const speedHorz = Math.random() * 10;
    const speedUp = Math.random() * 25;
    const spinVal = Math.random() * 360;
    const spinSpeed = Math.random() * 35 * (Math.random() <= 0.5 ? -1 : 1);
    const top = mouseY - size / 2;
    const left = mouseX - size / 2;
    const direction = Math.random() <= 0.5 ? -1 : 1;
    const particle = document.createElement("div");
    particle.innerHTML = `<img src="${imageUrl}" width="${size}" height="${size}" style="border-radius: 25%">`;
    particle.setAttribute("style", [
      "position:absolute",
      "will-change:transform",
      `top:${top}px`,
      `left:${left}px`,
      `transform:rotate(${spinVal}deg)`
    ].join(";"));
    container.appendChild(particle);
    particles.push({
      direction,
      element: particle,
      left,
      size,
      speedHorz,
      speedUp,
      spinSpeed,
      spinVal,
      top
    });
  }
  function updateParticles() {
    particles.forEach((p) => {
      p.left = p.left - p.speedHorz * p.direction;
      p.top = p.top - p.speedUp;
      p.speedUp = Math.min(p.size, p.speedUp - 1);
      p.spinVal = p.spinVal + p.spinSpeed;
      if (p.top >= Math.max(window.innerHeight, document.body.clientHeight) + p.size) {
        particles = particles.filter((o) => o !== p);
        p.element.remove();
      }
      p.element.setAttribute("style", [
        "position:absolute",
        "will-change:transform",
        `top:${p.top}px`,
        `left:${p.left}px`,
        `transform:rotate(${p.spinVal}deg)`
      ].join(";"));
    });
  }
  let animationFrame;
  function loop() {
    if (autoAddParticle && particles.length < limit) {
      createParticle();
    }
    updateParticles();
    animationFrame = requestAnimationFrame(loop);
  }
  loop();
  const isTouchInteraction = "ontouchstart" in window || navigator.msMaxTouchPoints;
  const tap = isTouchInteraction ? "touchstart" : "mousedown";
  const tapEnd = isTouchInteraction ? "touchend" : "mouseup";
  const move = isTouchInteraction ? "touchmove" : "mousemove";
  const updateMousePosition = (e) => {
    var _a, _b;
    if ("touches" in e) {
      mouseX = (_a = e.touches) == null ? void 0 : _a[0].clientX;
      mouseY = (_b = e.touches) == null ? void 0 : _b[0].clientY;
    } else {
      mouseX = e.clientX;
      mouseY = e.clientY;
    }
  };
  const tapHandler = (e) => {
    updateMousePosition(e);
    autoAddParticle = true;
  };
  const disableAutoAddParticle = () => {
    autoAddParticle = false;
  };
  element2.addEventListener(move, updateMousePosition, { passive: false });
  element2.addEventListener(tap, tapHandler);
  element2.addEventListener(tapEnd, disableAutoAddParticle);
  element2.addEventListener("mouseleave", disableAutoAddParticle);
  return () => {
    element2.removeEventListener(move, updateMousePosition);
    element2.removeEventListener(tap, tapHandler);
    element2.removeEventListener(tapEnd, disableAutoAddParticle);
    element2.removeEventListener("mouseleave", disableAutoAddParticle);
    let interval = setInterval(() => {
      if (animationFrame && particles.length === 0) {
        cancelAnimationFrame(animationFrame);
        clearInterval(interval);
        if (--instanceCounter === 0) {
          container.remove();
        }
      }
    }, 500);
  };
}

// src/components/ModalSelection/ModalSelection.css.ts
var HoverClassName = "g5kl0l1 ju367v6x ju367v7h ju367v81 ju367v6d ju367vah ju367v11 ju367v16 ju367v18 ju367v8 ju367vc";
var SelectedClassName = "ju367v6x ju367v7h ju367v81 ju367v6d ju367v94 ju367vdm ju367v11 ju367v16 ju367v18 ju367ve4 ju367vc";

// src/components/ModalSelection/ModalSelection.tsx
var ModalSelection = ({
  as = "button",
  currentlySelected = false,
  iconBackground,
  iconUrl,
  name,
  onClick,
  ready,
  ...urlProps
}) => {
  const coolModeRef = useCoolMode(iconUrl);
  const [isMouseOver, setIsMouseOver] = useState5(false);
  return /* @__PURE__ */ React32.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    onMouseEnter: () => setIsMouseOver(true),
    onMouseLeave: () => setIsMouseOver(false),
    ref: coolModeRef
  }, /* @__PURE__ */ React32.createElement(Box, {
    as,
    className: currentlySelected ? SelectedClassName : HoverClassName,
    disabled: currentlySelected,
    onClick,
    style: { willChange: "transform" },
    ...urlProps
  }, /* @__PURE__ */ React32.createElement(Box, {
    color: currentlySelected ? "accentColorForeground" : "modalText",
    disabled: !ready,
    fontFamily: "body",
    fontSize: "16",
    fontWeight: "bold",
    transition: "default"
  }, /* @__PURE__ */ React32.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "row",
    gap: "12"
  }, /* @__PURE__ */ React32.createElement(AsyncImage, {
    background: iconBackground,
    ...isMouseOver ? {} : { borderColor: "actionButtonBorder" },
    borderRadius: "6",
    height: "28",
    src: iconUrl,
    width: "28"
  }), /* @__PURE__ */ React32.createElement("div", null, name)))));
};
ModalSelection.displayName = "ModalSelection";

// src/components/ConnectOptions/ConnectDetails.tsx
import React36, { useEffect as useEffect11 } from "react";

// src/hooks/useWindowSize.ts
import { useEffect as useEffect10, useState as useState6 } from "react";
var useWindowSize = () => {
  const [windowSize, setWindowSize] = useState6({
    height: window.innerHeight,
    width: window.innerWidth
  });
  useEffect10(() => {
    function handleResize() {
      setWindowSize({
        height: window.innerHeight,
        width: window.innerWidth
      });
    }
    window.addEventListener("resize", handleResize);
    handleResize();
    return () => window.removeEventListener("resize", handleResize);
  }, []);
  return windowSize;
};

// src/components/Icons/Create.tsx
import React33 from "react";
var src3 = async () => (await import("./create-GVGWOEGN.js")).default;
var preloadCreateIcon = () => loadImages(src3);
var CreateIcon = () => /* @__PURE__ */ React33.createElement(AsyncImage, {
  background: "#e3a5e8",
  height: "48",
  src: src3,
  width: "48"
});

// src/components/Icons/Scan.tsx
import React34 from "react";
var src4 = async () => (await import("./scan-RESLZYEA.js")).default;
var preloadScanIcon = () => loadImages(src4);
var ScanIcon = () => /* @__PURE__ */ React34.createElement(AsyncImage, {
  background: "#515a70",
  height: "48",
  src: src4,
  width: "48"
});

// src/components/QRCode/QRCode.tsx
import QRCodeUtil from "qrcode";
import React35, { useMemo as useMemo5 } from "react";

// src/components/ConnectOptions/DesktopOptions.css.ts
var QRCodeBackgroundClassName = "_1vwt0cg0";
var ScrollClassName = "_1vwt0cg2 ju367v74 ju367v7o";
var sidebar = "_1vwt0cg3";

// src/components/QRCode/QRCode.tsx
var generateMatrix = (value, errorCorrectionLevel) => {
  const arr = Array.prototype.slice.call(QRCodeUtil.create(value, { errorCorrectionLevel }).modules.data, 0);
  const sqrt = Math.sqrt(arr.length);
  return arr.reduce((rows, key, index) => (index % sqrt === 0 ? rows.push([key]) : rows[rows.length - 1].push(key)) && rows, []);
};
function QRCode({
  ecl = "M",
  logoBackground,
  logoMargin = 10,
  logoSize = 50,
  logoUrl,
  size: sizeProp = 200,
  uri
}) {
  const padding = "20";
  const size = sizeProp - parseInt(padding, 10) * 2;
  const dots = useMemo5(() => {
    const dots2 = [];
    const matrix = generateMatrix(uri, ecl);
    const cellSize = size / matrix.length;
    let qrList = [
      { x: 0, y: 0 },
      { x: 1, y: 0 },
      { x: 0, y: 1 }
    ];
    qrList.forEach(({ x, y }) => {
      const x1 = (matrix.length - 7) * cellSize * x;
      const y1 = (matrix.length - 7) * cellSize * y;
      for (let i = 0; i < 3; i++) {
        dots2.push(/* @__PURE__ */ React35.createElement("rect", {
          fill: i % 2 !== 0 ? "white" : "black",
          height: cellSize * (7 - i * 2),
          key: `${i}-${x}-${y}`,
          rx: (i - 2) * -5 + (i === 0 ? 2 : 0),
          ry: (i - 2) * -5 + (i === 0 ? 2 : 0),
          width: cellSize * (7 - i * 2),
          x: x1 + cellSize * i,
          y: y1 + cellSize * i
        }));
      }
    });
    const clearArenaSize = Math.floor((logoSize + 25) / cellSize);
    const matrixMiddleStart = matrix.length / 2 - clearArenaSize / 2;
    const matrixMiddleEnd = matrix.length / 2 + clearArenaSize / 2 - 1;
    matrix.forEach((row, i) => {
      row.forEach((_, j) => {
        if (matrix[i][j]) {
          if (!(i < 7 && j < 7 || i > matrix.length - 8 && j < 7 || i < 7 && j > matrix.length - 8)) {
            if (!(i > matrixMiddleStart && i < matrixMiddleEnd && j > matrixMiddleStart && j < matrixMiddleEnd)) {
              dots2.push(/* @__PURE__ */ React35.createElement("circle", {
                cx: i * cellSize + cellSize / 2,
                cy: j * cellSize + cellSize / 2,
                fill: "black",
                key: `circle-${i}-${j}`,
                r: cellSize / 3
              }));
            }
          }
        }
      });
    });
    return dots2;
  }, [ecl, logoSize, size, uri]);
  const logoPosition = size / 2 - logoSize / 2;
  const logoWrapperSize = logoSize + logoMargin * 2;
  return /* @__PURE__ */ React35.createElement(Box, {
    borderColor: "generalBorder",
    borderRadius: "menuButton",
    borderStyle: "solid",
    borderWidth: "1",
    className: QRCodeBackgroundClassName,
    padding,
    width: "max"
  }, /* @__PURE__ */ React35.createElement(Box, {
    style: {
      height: size,
      userSelect: "none",
      width: size
    },
    userSelect: "none"
  }, /* @__PURE__ */ React35.createElement(Box, {
    display: "flex",
    justifyContent: "center",
    position: "relative",
    style: {
      height: 0,
      top: logoPosition,
      width: size
    },
    width: "full"
  }, /* @__PURE__ */ React35.createElement(AsyncImage, {
    background: logoBackground,
    borderColor: { custom: "rgba(0, 0, 0, 0.06)" },
    borderRadius: "13",
    height: logoSize,
    src: logoUrl,
    width: logoSize
  })), /* @__PURE__ */ React35.createElement("svg", {
    height: size,
    style: { all: "revert" },
    width: size
  }, /* @__PURE__ */ React35.createElement("defs", null, /* @__PURE__ */ React35.createElement("clipPath", {
    id: "clip-wrapper"
  }, /* @__PURE__ */ React35.createElement("rect", {
    height: logoWrapperSize,
    width: logoWrapperSize
  })), /* @__PURE__ */ React35.createElement("clipPath", {
    id: "clip-logo"
  }, /* @__PURE__ */ React35.createElement("rect", {
    height: logoSize,
    width: logoSize
  }))), /* @__PURE__ */ React35.createElement("rect", {
    fill: "transparent",
    height: size,
    width: size
  }), dots)));
}

// src/components/ConnectOptions/ConnectDetails.tsx
function GetDetail({
  getMobileWallet
}) {
  const wallets = useWalletConnectors();
  const shownWallets = wallets.splice(0, 5);
  const linkProps = (link) => ({
    as: "a",
    href: link,
    rel: "noreferrer",
    target: "_blank"
  });
  return /* @__PURE__ */ React36.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    height: "full",
    marginTop: "18",
    width: "full"
  }, /* @__PURE__ */ React36.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    gap: "28",
    height: "full",
    width: "full"
  }, shownWallets == null ? void 0 : shownWallets.filter((wallet) => {
    var _a, _b;
    return ((_a = wallet.downloadUrls) == null ? void 0 : _a.browserExtension) || wallet.qrCode && ((_b = wallet.downloadUrls) == null ? void 0 : _b.qrCode);
  }).map((wallet) => {
    const { downloadUrls, iconBackground, iconUrl, id, name, qrCode } = wallet;
    const hasMobileCompanionApp = (downloadUrls == null ? void 0 : downloadUrls.qrCode) && qrCode;
    return /* @__PURE__ */ React36.createElement(Box, {
      alignItems: "center",
      display: "flex",
      gap: "16",
      justifyContent: "space-between",
      key: wallet.id,
      width: "full"
    }, /* @__PURE__ */ React36.createElement(Box, {
      alignItems: "center",
      display: "flex",
      flexDirection: "row",
      gap: "16"
    }, /* @__PURE__ */ React36.createElement(AsyncImage, {
      background: iconBackground,
      borderColor: "actionButtonBorder",
      borderRadius: "10",
      height: "48",
      src: iconUrl,
      width: "48"
    }), /* @__PURE__ */ React36.createElement(Box, {
      display: "flex",
      flexDirection: "column",
      gap: "2"
    }, /* @__PURE__ */ React36.createElement(Text, {
      color: "modalText",
      size: "14",
      weight: "bold"
    }, name), /* @__PURE__ */ React36.createElement(Text, {
      color: "modalTextSecondary",
      size: "14",
      weight: "medium"
    }, hasMobileCompanionApp ? "Mobile Wallet" : (downloadUrls == null ? void 0 : downloadUrls.browserExtension) ? "Browser Extension" : null))), /* @__PURE__ */ React36.createElement(Box, {
      display: "flex",
      flexDirection: "column",
      gap: "4",
      onClick: () => hasMobileCompanionApp && getMobileWallet(id),
      ...!hasMobileCompanionApp && (downloadUrls == null ? void 0 : downloadUrls.browserExtension) ? linkProps(downloadUrls.browserExtension) : {}
    }, /* @__PURE__ */ React36.createElement(ActionButton, {
      label: "GET",
      onClick: () => {
      },
      type: "secondary"
    })));
  })), /* @__PURE__ */ React36.createElement(Box, {
    alignItems: "center",
    borderRadius: "10",
    display: "flex",
    flexDirection: "column",
    gap: "8",
    justifyContent: "space-between",
    marginBottom: "4",
    paddingY: "8",
    style: { maxWidth: 275, textAlign: "center" }
  }, /* @__PURE__ */ React36.createElement(Text, {
    color: "modalText",
    size: "14",
    weight: "bold"
  }, "Not what you\u2019re looking for?"), /* @__PURE__ */ React36.createElement(Text, {
    color: "modalTextSecondary",
    size: "14",
    weight: "medium"
  }, "Select a wallet on the left to get started with a different wallet provider.")));
}
var LOGO_SIZE = "60";
function ConnectDetail({
  connectionError,
  qrCodeUri,
  reconnect,
  setWalletStep,
  wallet
}) {
  var _a;
  const {
    downloadUrls,
    iconBackground,
    iconUrl,
    name,
    qrCode,
    ready,
    showWalletConnectModal
  } = wallet;
  const getDesktopDeepLink = (_a = wallet.desktop) == null ? void 0 : _a.getUri;
  const safari = isSafari();
  let readyMsg;
  if (ready) {
    readyMsg = "Waiting for connection";
  } else if (downloadUrls == null ? void 0 : downloadUrls.browserExtension) {
    readyMsg = `The ${name} extension is not installed in your browser`;
  } else {
    readyMsg = `${name} is not available on this device`;
  }
  const secondaryAction = showWalletConnectModal ? {
    description: "Need the official WalletConnect modal?",
    label: "OPEN",
    onClick: showWalletConnectModal
  } : qrCode ? {
    description: `Don\u2019t have the ${name} app?`,
    label: "GET",
    onClick: () => setWalletStep("DOWNLOAD" /* Download */)
  } : {
    description: `Confirm the connection in ${name}`,
    label: "RETRY",
    onClick: getDesktopDeepLink ? async () => {
      const uri = await getDesktopDeepLink();
      window.open(uri, safari ? "_blank" : "_self");
    } : () => reconnect(wallet)
  };
  const { width: windowWidth } = useWindowSize();
  const smallWindow = windowWidth && windowWidth < 768;
  return /* @__PURE__ */ React36.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    height: "full",
    width: "full"
  }, qrCode && qrCodeUri ? /* @__PURE__ */ React36.createElement(Box, {
    alignItems: "center",
    display: "flex",
    height: "full",
    justifyContent: "center"
  }, /* @__PURE__ */ React36.createElement(QRCode, {
    logoBackground: iconBackground,
    logoSize: smallWindow ? 60 : 72,
    logoUrl: iconUrl,
    size: smallWindow ? Math.max(280, Math.min(windowWidth - 308, 382)) : 382,
    uri: qrCodeUri
  })) : /* @__PURE__ */ React36.createElement(Box, {
    alignItems: "center",
    display: "flex",
    justifyContent: "center",
    style: { flexGrow: 1 }
  }, /* @__PURE__ */ React36.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    gap: "20"
  }, /* @__PURE__ */ React36.createElement(Box, {
    borderRadius: "10",
    height: LOGO_SIZE,
    overflow: "hidden"
  }, /* @__PURE__ */ React36.createElement(AsyncImage, {
    height: LOGO_SIZE,
    src: iconUrl,
    width: LOGO_SIZE
  })), /* @__PURE__ */ React36.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    gap: "10",
    paddingX: "32",
    style: { textAlign: "center" }
  }, /* @__PURE__ */ React36.createElement(Text, {
    color: "modalText",
    size: "20",
    weight: "bold"
  }, ready ? `Opening ${name}` : (downloadUrls == null ? void 0 : downloadUrls.browserExtension) ? `${name} is not installed` : `${name} is not available`), /* @__PURE__ */ React36.createElement(Box, {
    color: "modalTextSecondary",
    display: "flex",
    flexDirection: "row",
    gap: "6",
    height: "24"
  }, connectionError ? /* @__PURE__ */ React36.createElement(Text, {
    color: "error",
    size: "16",
    textAlign: "center",
    weight: "bold"
  }, "Error connecting, please retry!") : /* @__PURE__ */ React36.createElement(React36.Fragment, null, ready ? /* @__PURE__ */ React36.createElement(SpinnerIcon, null) : null, /* @__PURE__ */ React36.createElement(Text, {
    color: "modalTextSecondary",
    size: "16",
    textAlign: "center",
    weight: "bold"
  }, readyMsg))), !ready && (downloadUrls == null ? void 0 : downloadUrls.browserExtension) ? /* @__PURE__ */ React36.createElement(Box, {
    paddingTop: "20"
  }, /* @__PURE__ */ React36.createElement(ActionButton, {
    href: downloadUrls.browserExtension,
    label: "INSTALL",
    type: "secondary"
  })) : null))), /* @__PURE__ */ React36.createElement(Box, {
    alignItems: "center",
    borderRadius: "10",
    display: "flex",
    flexDirection: "row",
    gap: "8",
    height: "28",
    justifyContent: "space-between",
    marginTop: "6"
  }, !ready ? null : /* @__PURE__ */ React36.createElement(React36.Fragment, null, /* @__PURE__ */ React36.createElement(Text, {
    color: "modalTextSecondary",
    size: "14",
    weight: "medium"
  }, secondaryAction.description), /* @__PURE__ */ React36.createElement(ActionButton, {
    label: secondaryAction.label,
    onClick: secondaryAction.onClick,
    type: "secondary"
  }))));
}
function DownloadDetail({
  setWalletStep,
  wallet
}) {
  const { downloadUrls, qrCode } = wallet;
  useEffect11(() => {
    preloadCreateIcon();
    preloadScanIcon();
  }, []);
  return /* @__PURE__ */ React36.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    gap: "24",
    height: "full",
    width: "full"
  }, /* @__PURE__ */ React36.createElement(Box, {
    style: { maxWidth: 220, textAlign: "center" }
  }, /* @__PURE__ */ React36.createElement(Text, {
    color: "modalTextSecondary",
    size: "14",
    weight: "semibold"
  }, "Scan with your phone to download on iOS or Android")), /* @__PURE__ */ React36.createElement(Box, {
    height: "full"
  }, (downloadUrls == null ? void 0 : downloadUrls.qrCode) ? /* @__PURE__ */ React36.createElement(QRCode, {
    logoSize: 0,
    size: 268,
    uri: downloadUrls.qrCode
  }) : null), /* @__PURE__ */ React36.createElement(Box, {
    alignItems: "center",
    borderRadius: "10",
    display: "flex",
    flexDirection: "row",
    gap: "8",
    height: "34",
    justifyContent: "space-between",
    marginBottom: "12",
    paddingY: "8"
  }, /* @__PURE__ */ React36.createElement(ActionButton, {
    label: "Continue",
    onClick: () => setWalletStep((qrCode == null ? void 0 : qrCode.instructions) ? "INSTRUCTIONS" /* Instructions */ : "CONNECT" /* Connect */)
  })));
}
var stepIcons = {
  create: () => /* @__PURE__ */ React36.createElement(CreateIcon, null),
  install: (wallet) => /* @__PURE__ */ React36.createElement(AsyncImage, {
    background: wallet.iconBackground,
    borderColor: "generalBorder",
    borderRadius: "10",
    height: "48",
    src: wallet.iconUrl,
    width: "48"
  }),
  scan: () => /* @__PURE__ */ React36.createElement(ScanIcon, null)
};
function InstructionDetail({
  setWalletStep,
  wallet
}) {
  var _a, _b, _c, _d;
  return /* @__PURE__ */ React36.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    height: "full",
    width: "full"
  }, /* @__PURE__ */ React36.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    gap: "28",
    height: "full",
    justifyContent: "center",
    paddingY: "32",
    style: { maxWidth: 320 }
  }, (_b = (_a = wallet == null ? void 0 : wallet.qrCode) == null ? void 0 : _a.instructions) == null ? void 0 : _b.steps.map((d, idx) => {
    var _a2;
    return /* @__PURE__ */ React36.createElement(Box, {
      alignItems: "center",
      display: "flex",
      flexDirection: "row",
      gap: "16",
      key: idx
    }, /* @__PURE__ */ React36.createElement(Box, {
      borderRadius: "10",
      height: "48",
      minWidth: "48",
      overflow: "hidden",
      width: "48"
    }, (_a2 = stepIcons[d.step]) == null ? void 0 : _a2.call(stepIcons, wallet)), /* @__PURE__ */ React36.createElement(Box, {
      display: "flex",
      flexDirection: "column",
      gap: "4"
    }, /* @__PURE__ */ React36.createElement(Text, {
      color: "modalText",
      size: "14",
      weight: "bold"
    }, d.title), /* @__PURE__ */ React36.createElement(Text, {
      color: "modalTextSecondary",
      size: "14",
      weight: "medium"
    }, d.description)));
  })), /* @__PURE__ */ React36.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    gap: "12",
    justifyContent: "center",
    marginBottom: "16"
  }, /* @__PURE__ */ React36.createElement(ActionButton, {
    label: "Connect",
    onClick: () => setWalletStep("CONNECT" /* Connect */)
  }), /* @__PURE__ */ React36.createElement(Box, {
    as: "a",
    href: (_d = (_c = wallet == null ? void 0 : wallet.qrCode) == null ? void 0 : _c.instructions) == null ? void 0 : _d.learnMoreUrl,
    paddingX: "12",
    paddingY: "4",
    rel: "noreferrer",
    style: { willChange: "transform" },
    target: "_blank",
    transform: { active: "shrink", hover: "grow" },
    transition: "default"
  }, /* @__PURE__ */ React36.createElement(Text, {
    color: "accentColor",
    size: "14",
    weight: "bold"
  }, "Learn More"))));
}

// src/components/ConnectOptions/DesktopOptions.tsx
function DesktopOptions({ onClose }) {
  const titleId = "rk_connect_title";
  const safari = isSafari();
  const [selectedOptionId, setSelectedOptionId] = useState7();
  const [selectedWallet, setSelectedWallet] = useState7();
  const [qrCodeUri, setQrCodeUri] = useState7();
  const hasQrCode = !!(selectedWallet == null ? void 0 : selectedWallet.qrCode);
  const [connectionError, setConnectionError] = useState7(false);
  const wallets = useWalletConnectors().filter((wallet) => {
    var _a;
    return wallet.ready || ((_a = wallet.downloadUrls) == null ? void 0 : _a.browserExtension);
  });
  const groupedWallets = groupBy(wallets, (wallet) => wallet.groupName);
  const connectToWallet = (wallet) => {
    var _a, _b, _c;
    setConnectionError(false);
    if (wallet.ready) {
      (_b = (_a = wallet == null ? void 0 : wallet.connect) == null ? void 0 : _a.call(wallet)) == null ? void 0 : _b.catch(() => {
        setConnectionError(true);
      });
      const getDesktopDeepLink = (_c = wallet.desktop) == null ? void 0 : _c.getUri;
      if (getDesktopDeepLink) {
        setTimeout(async () => {
          const uri = await getDesktopDeepLink();
          window.open(uri, safari ? "_blank" : "_self");
        }, 0);
      }
    }
  };
  const onSelectWallet = (wallet) => {
    var _a;
    connectToWallet(wallet);
    setSelectedOptionId(wallet.id);
    if (wallet.ready) {
      (_a = wallet == null ? void 0 : wallet.onConnecting) == null ? void 0 : _a.call(wallet, async () => {
        var _a2;
        const sWallet = wallets.find((w) => wallet.id === w.id);
        const uri = await ((_a2 = sWallet == null ? void 0 : sWallet.qrCode) == null ? void 0 : _a2.getUri());
        setQrCodeUri(uri);
        setSelectedWallet(sWallet);
        setWalletStep("CONNECT" /* Connect */);
      });
    } else {
      setSelectedWallet(wallet);
      setWalletStep("CONNECT" /* Connect */);
    }
  };
  const getMobileWallet = (id) => {
    setSelectedOptionId(id);
    const sWallet = wallets.find((w) => id === w.id);
    setSelectedWallet(sWallet);
    setWalletStep("DOWNLOAD" /* Download */);
  };
  const [walletStep, setWalletStep] = useState7("NONE" /* None */);
  let walletContent = null;
  let headerLabel = null;
  let headerBackButtonLink = null;
  useEffect12(() => {
    setConnectionError(false);
  }, [walletStep, selectedWallet]);
  switch (walletStep) {
    case "NONE" /* None */:
      walletContent = /* @__PURE__ */ React37.createElement(ConnectModalIntro, {
        getWallet: () => setWalletStep("GET" /* Get */)
      });
      break;
    case "GET" /* Get */:
      walletContent = /* @__PURE__ */ React37.createElement(GetDetail, {
        getMobileWallet
      });
      headerLabel = "Get a Wallet";
      headerBackButtonLink = "NONE" /* None */;
      break;
    case "CONNECT" /* Connect */:
      walletContent = selectedWallet && /* @__PURE__ */ React37.createElement(ConnectDetail, {
        connectionError,
        qrCodeUri,
        reconnect: connectToWallet,
        setWalletStep,
        wallet: selectedWallet
      });
      headerLabel = hasQrCode && `Scan with ${selectedWallet.name === "WalletConnect" ? "your phone" : selectedWallet.name}`;
      break;
    case "DOWNLOAD" /* Download */:
      walletContent = selectedWallet && /* @__PURE__ */ React37.createElement(DownloadDetail, {
        setWalletStep,
        wallet: selectedWallet
      });
      headerLabel = hasQrCode && `Install ${selectedWallet.name}`;
      headerBackButtonLink = "CONNECT" /* Connect */;
      break;
    case "INSTRUCTIONS" /* Instructions */:
      walletContent = selectedWallet && /* @__PURE__ */ React37.createElement(InstructionDetail, {
        setWalletStep,
        wallet: selectedWallet
      });
      headerLabel = hasQrCode && `Get started with ${selectedWallet.name}`;
      headerBackButtonLink = "DOWNLOAD" /* Download */;
      break;
    default:
      break;
  }
  return /* @__PURE__ */ React37.createElement(Box, {
    display: "flex",
    flexDirection: "row"
  }, /* @__PURE__ */ React37.createElement(Box, {
    className: sidebar,
    display: "flex",
    flexDirection: "column",
    marginTop: "18"
  }, /* @__PURE__ */ React37.createElement(Box, {
    marginLeft: "6",
    paddingBottom: "10",
    paddingX: "18"
  }, /* @__PURE__ */ React37.createElement(Text, {
    as: "h1",
    color: "modalText",
    id: titleId,
    size: "18",
    weight: "heavy"
  }, "Connect a Wallet")), /* @__PURE__ */ React37.createElement(Box, {
    className: ScrollClassName,
    paddingBottom: "18"
  }, Object.entries(groupedWallets).map(([groupName, wallets2], index) => wallets2.length > 0 && /* @__PURE__ */ React37.createElement(Fragment2, {
    key: index
  }, groupName ? /* @__PURE__ */ React37.createElement(Box, {
    marginBottom: "8",
    marginTop: "16",
    marginX: "6"
  }, /* @__PURE__ */ React37.createElement(Text, {
    color: "modalTextSecondary",
    size: "14",
    weight: "bold"
  }, groupName)) : null, /* @__PURE__ */ React37.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    gap: "4"
  }, wallets2.map((wallet) => {
    return /* @__PURE__ */ React37.createElement(ModalSelection, {
      currentlySelected: wallet.id === selectedOptionId,
      iconBackground: wallet.iconBackground,
      iconUrl: wallet.iconUrl,
      key: wallet.id,
      name: wallet.name,
      onClick: () => onSelectWallet(wallet),
      ready: wallet.ready
    });
  })))))), !isMobile() && /* @__PURE__ */ React37.createElement(React37.Fragment, null, /* @__PURE__ */ React37.createElement(Box, {
    background: "generalBorder",
    minWidth: "1",
    width: "1"
  }), /* @__PURE__ */ React37.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    margin: "16",
    style: { flexGrow: 1 }
  }, /* @__PURE__ */ React37.createElement(Box, {
    alignItems: "center",
    display: "flex",
    justifyContent: "space-between",
    marginBottom: "12"
  }, /* @__PURE__ */ React37.createElement(Box, {
    width: "28"
  }, headerBackButtonLink && /* @__PURE__ */ React37.createElement(Box, {
    as: "button",
    color: "accentColor",
    onClick: () => headerBackButtonLink && setWalletStep(headerBackButtonLink),
    paddingX: "8",
    paddingY: "4",
    style: {
      boxSizing: "content-box",
      height: 17,
      willChange: "transform"
    },
    transform: { active: "shrinkSm", hover: "growLg" },
    transition: "default"
  }, /* @__PURE__ */ React37.createElement(BackIcon, null))), /* @__PURE__ */ React37.createElement(Box, {
    display: "flex",
    justifyContent: "center",
    style: { flexGrow: 1 }
  }, headerLabel && /* @__PURE__ */ React37.createElement(Text, {
    color: "modalText",
    size: "18",
    textAlign: "center",
    weight: "heavy"
  }, headerLabel)), /* @__PURE__ */ React37.createElement(CloseButton, {
    onClose
  })), /* @__PURE__ */ React37.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    style: { minHeight: 432 }
  }, /* @__PURE__ */ React37.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    gap: "6",
    height: "full",
    justifyContent: "center",
    marginTop: "6",
    marginX: "8"
  }, walletContent)))));
}

// src/components/ConnectOptions/MobileOptions.tsx
import React38, { useCallback as useCallback6, useContext as useContext9, useState as useState8 } from "react";

// src/components/ConnectOptions/MobileOptions.css.ts
var scroll = "_1am14410";

// src/components/ConnectOptions/MobileOptions.tsx
function WalletButton({ wallet }) {
  const {
    connect,
    iconBackground,
    iconUrl,
    id,
    mobile,
    name,
    onConnecting,
    ready,
    shortName
  } = wallet;
  const getMobileUri = mobile == null ? void 0 : mobile.getUri;
  const coolModeRef = useCoolMode(iconUrl);
  return /* @__PURE__ */ React38.createElement(Box, {
    as: "button",
    color: ready ? "modalText" : "modalTextSecondary",
    disabled: !ready,
    fontFamily: "body",
    key: id,
    onClick: useCallback6(async () => {
      connect == null ? void 0 : connect();
      onConnecting == null ? void 0 : onConnecting(async () => {
        if (getMobileUri) {
          window.location.href = await getMobileUri();
        }
      });
    }, [connect, getMobileUri, onConnecting]),
    ref: coolModeRef,
    style: { overflow: "visible", textAlign: "center" },
    type: "button",
    width: "full"
  }, /* @__PURE__ */ React38.createElement(Box, {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    justifyContent: "center"
  }, /* @__PURE__ */ React38.createElement(Box, {
    paddingBottom: "8",
    paddingTop: "10"
  }, /* @__PURE__ */ React38.createElement(AsyncImage, {
    background: iconBackground,
    borderRadius: "13",
    boxShadow: "walletLogo",
    height: "60",
    src: iconUrl,
    width: "60"
  })), /* @__PURE__ */ React38.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    gap: "1",
    textAlign: "center"
  }, /* @__PURE__ */ React38.createElement(Text, {
    as: "h2",
    color: wallet.ready ? "modalText" : "modalTextSecondary",
    size: "13",
    weight: "medium"
  }, /* @__PURE__ */ React38.createElement(Box, {
    as: "span",
    position: "relative"
  }, shortName != null ? shortName : name, !wallet.ready && " (unsupported)")), wallet.recent && /* @__PURE__ */ React38.createElement(Text, {
    color: "accentColor",
    size: "12",
    weight: "medium"
  }, "Recent"))));
}
function MobileOptions({ onClose }) {
  var _a;
  const titleId = "rk_connect_title";
  const wallets = useWalletConnectors();
  const { learnMoreUrl } = useContext9(AppContext);
  let headerLabel = null;
  let walletContent = null;
  let headerBackgroundContrast = false;
  let headerBackButtonLink = null;
  const [walletStep, setWalletStep] = useState8("CONNECT" /* Connect */);
  const ios = isIOS();
  switch (walletStep) {
    case "CONNECT" /* Connect */: {
      headerLabel = "Connect a Wallet";
      headerBackgroundContrast = true;
      walletContent = /* @__PURE__ */ React38.createElement(React38.Fragment, null, /* @__PURE__ */ React38.createElement(Box, {
        background: "profileForeground",
        className: scroll,
        display: "flex",
        paddingBottom: "20",
        paddingTop: "6"
      }, /* @__PURE__ */ React38.createElement(Box, {
        display: "flex",
        style: { margin: "0 auto" }
      }, wallets.filter((wallet) => wallet.ready).map((wallet) => {
        return /* @__PURE__ */ React38.createElement(Box, {
          key: wallet.id,
          paddingX: "20"
        }, /* @__PURE__ */ React38.createElement(Box, {
          width: "60"
        }, /* @__PURE__ */ React38.createElement(WalletButton, {
          wallet
        })));
      }))), /* @__PURE__ */ React38.createElement(Box, {
        background: "generalBorder",
        height: "1",
        marginBottom: "32",
        marginTop: "-1"
      }), /* @__PURE__ */ React38.createElement(Box, {
        alignItems: "center",
        display: "flex",
        flexDirection: "column",
        gap: "32",
        paddingX: "32",
        style: { textAlign: "center" }
      }, /* @__PURE__ */ React38.createElement(Box, {
        display: "flex",
        flexDirection: "column",
        gap: "8",
        textAlign: "center"
      }, /* @__PURE__ */ React38.createElement(Text, {
        color: "modalText",
        size: "16",
        weight: "bold"
      }, "What is a Wallet?"), /* @__PURE__ */ React38.createElement(Text, {
        color: "modalTextSecondary",
        size: "16"
      }, "A wallet is used to send, receive, store, and display digital assets. It\u2019s also a new way to log in, without needing to create new accounts and passwords on\xA0every\xA0website."))), /* @__PURE__ */ React38.createElement(Box, {
        paddingTop: "32",
        paddingX: "20"
      }, /* @__PURE__ */ React38.createElement(Box, {
        display: "flex",
        gap: "14",
        justifyContent: "center"
      }, /* @__PURE__ */ React38.createElement(ActionButton, {
        label: "Get a Wallet",
        onClick: () => setWalletStep("GET" /* Get */),
        size: "large",
        type: "secondary"
      }), /* @__PURE__ */ React38.createElement(ActionButton, {
        href: learnMoreUrl,
        label: "Learn More",
        size: "large",
        type: "secondary"
      }))));
      break;
    }
    case "GET" /* Get */: {
      headerLabel = "Get a Wallet";
      headerBackButtonLink = "CONNECT" /* Connect */;
      const mobileWallets = (_a = wallets == null ? void 0 : wallets.filter((wallet) => {
        var _a2, _b;
        return ((_a2 = wallet.downloadUrls) == null ? void 0 : _a2.ios) || ((_b = wallet.downloadUrls) == null ? void 0 : _b.android);
      })) == null ? void 0 : _a.splice(0, 3);
      walletContent = /* @__PURE__ */ React38.createElement(React38.Fragment, null, /* @__PURE__ */ React38.createElement(Box, {
        alignItems: "center",
        display: "flex",
        flexDirection: "column",
        height: "full",
        marginBottom: "36",
        marginTop: "5",
        paddingTop: "12",
        width: "full"
      }, mobileWallets.map((wallet, index) => {
        const { downloadUrls, iconBackground, iconUrl, name } = wallet;
        if (!(downloadUrls == null ? void 0 : downloadUrls.ios) && !(downloadUrls == null ? void 0 : downloadUrls.android)) {
          return null;
        }
        return /* @__PURE__ */ React38.createElement(Box, {
          display: "flex",
          gap: "16",
          key: wallet.id,
          paddingX: "20",
          width: "full"
        }, /* @__PURE__ */ React38.createElement(Box, {
          style: { minHeight: 48, minWidth: 48 }
        }, /* @__PURE__ */ React38.createElement(AsyncImage, {
          background: iconBackground,
          borderColor: "generalBorder",
          borderRadius: "10",
          height: "48",
          src: iconUrl,
          width: "48"
        })), /* @__PURE__ */ React38.createElement(Box, {
          display: "flex",
          flexDirection: "column",
          width: "full"
        }, /* @__PURE__ */ React38.createElement(Box, {
          alignItems: "center",
          display: "flex",
          height: "48"
        }, /* @__PURE__ */ React38.createElement(Box, {
          width: "full"
        }, /* @__PURE__ */ React38.createElement(Text, {
          color: "modalText",
          size: "18",
          weight: "bold"
        }, name)), /* @__PURE__ */ React38.createElement(ActionButton, {
          href: ios ? downloadUrls == null ? void 0 : downloadUrls.ios : downloadUrls == null ? void 0 : downloadUrls.android,
          label: "GET",
          size: "small",
          type: "secondary"
        })), index < mobileWallets.length - 1 && /* @__PURE__ */ React38.createElement(Box, {
          background: "generalBorderDim",
          height: "1",
          marginY: "10",
          width: "full"
        })));
      })), /* @__PURE__ */ React38.createElement(Box, {
        style: { marginBottom: "42px" }
      }), /* @__PURE__ */ React38.createElement(Box, {
        alignItems: "center",
        display: "flex",
        flexDirection: "column",
        gap: "36",
        paddingX: "36",
        style: { textAlign: "center" }
      }, /* @__PURE__ */ React38.createElement(Box, {
        display: "flex",
        flexDirection: "column",
        gap: "12",
        textAlign: "center"
      }, /* @__PURE__ */ React38.createElement(Text, {
        color: "modalText",
        size: "16",
        weight: "bold"
      }, "Not what you\u2019re looking for?"), /* @__PURE__ */ React38.createElement(Text, {
        color: "modalTextSecondary",
        size: "16"
      }, "Select a wallet on the main screen to get started with a different wallet provider."))));
      break;
    }
  }
  return /* @__PURE__ */ React38.createElement(Box, {
    display: "flex",
    flexDirection: "column",
    paddingBottom: "36"
  }, /* @__PURE__ */ React38.createElement(Box, {
    background: headerBackgroundContrast ? "profileForeground" : "modalBackground",
    display: "flex",
    flexDirection: "column",
    paddingBottom: "4",
    paddingTop: "14"
  }, /* @__PURE__ */ React38.createElement(Box, {
    display: "flex",
    justifyContent: "center",
    paddingBottom: "6",
    paddingX: "20",
    position: "relative"
  }, headerBackButtonLink && /* @__PURE__ */ React38.createElement(Box, {
    display: "flex",
    position: "absolute",
    style: {
      left: 0,
      marginBottom: -20,
      marginTop: -20
    }
  }, /* @__PURE__ */ React38.createElement(Box, {
    alignItems: "center",
    as: "button",
    color: "accentColor",
    display: "flex",
    marginLeft: "4",
    marginTop: "20",
    onClick: () => setWalletStep(headerBackButtonLink),
    padding: "16",
    style: { height: 17, willChange: "transform" },
    transform: { active: "shrinkSm", hover: "growLg" },
    transition: "default"
  }, /* @__PURE__ */ React38.createElement(BackIcon, null))), /* @__PURE__ */ React38.createElement(Box, {
    marginTop: "4",
    textAlign: "center",
    width: "full"
  }, /* @__PURE__ */ React38.createElement(Text, {
    as: "h1",
    color: "modalText",
    id: titleId,
    size: "20",
    weight: "bold"
  }, headerLabel)), /* @__PURE__ */ React38.createElement(Box, {
    alignItems: "center",
    display: "flex",
    height: "32",
    paddingRight: "14",
    position: "absolute",
    right: "0"
  }, /* @__PURE__ */ React38.createElement(Box, {
    style: { marginBottom: -20, marginTop: -20 }
  }, /* @__PURE__ */ React38.createElement(CloseButton, {
    onClose
  }))))), /* @__PURE__ */ React38.createElement(Box, {
    display: "flex",
    flexDirection: "column"
  }, walletContent));
}

// src/components/ConnectOptions/ConnectOptions.tsx
function ConnectOptions({ onClose }) {
  return isMobile() ? /* @__PURE__ */ React39.createElement(MobileOptions, {
    onClose
  }) : /* @__PURE__ */ React39.createElement(DesktopOptions, {
    onClose
  });
}

// src/components/ConnectModal/ConnectModal.tsx
function ConnectModal({ onClose, open }) {
  const titleId = "rk_connect_title";
  const { data: connectData } = useAccount3();
  return !connectData ? /* @__PURE__ */ React40.createElement(Dialog, {
    onClose,
    open,
    titleId
  }, /* @__PURE__ */ React40.createElement(DialogContent, {
    bottomSheetOnMobile: true,
    padding: "0",
    wide: true
  }, /* @__PURE__ */ React40.createElement(ConnectOptions, {
    onClose
  }))) : null;
}

// src/components/ConnectButton/ConnectButtonRenderer.tsx
var useBooleanState = (initialValue) => {
  const [value, setValue] = useState9(initialValue);
  const setTrue = useCallback7(() => setValue(true), []);
  const setFalse = useCallback7(() => setValue(false), []);
  return { setFalse, setTrue, value };
};
function ConnectButtonRenderer({
  children
}) {
  var _a, _b;
  const mounted = useIsMounted();
  const { data: accountData } = useAccount4();
  const { data: ensAvatar } = useEnsAvatar({
    addressOrName: accountData == null ? void 0 : accountData.address
  });
  const { data: ensName } = useEnsName({ address: accountData == null ? void 0 : accountData.address });
  const { data: balanceData } = useBalance({
    addressOrName: accountData == null ? void 0 : accountData.address
  });
  const {
    activeChain,
    chains,
    error: networkError,
    switchNetwork
  } = useNetwork5();
  const { disconnect } = useDisconnect();
  const rainbowKitChains = useRainbowKitChains();
  const rainbowkitChainsById = useRainbowKitChainsById();
  const rainbowKitChain = activeChain ? rainbowkitChainsById[activeChain.id] : void 0;
  const chainIconUrl = (_a = rainbowKitChain == null ? void 0 : rainbowKitChain.iconUrl) != null ? _a : void 0;
  const chainIconBackground = (_b = rainbowKitChain == null ? void 0 : rainbowKitChain.iconBackground) != null ? _b : void 0;
  const resolvedChainIconUrl = useAsyncImage(chainIconUrl);
  const showRecentTransactions = useContext10(ShowRecentTransactionsContext);
  const hasPendingTransactions = useRecentTransactions().some(({ status }) => status === "pending") && showRecentTransactions;
  const {
    setFalse: closeConnectModal,
    setTrue: openConnectModal,
    value: connectModalOpen
  } = useBooleanState(false);
  const {
    setFalse: closeAccountModal,
    setTrue: openAccountModal,
    value: accountModalOpen
  } = useBooleanState(false);
  const {
    setFalse: closeChainModal,
    setTrue: openChainModal,
    value: chainModalOpen
  } = useBooleanState(false);
  const hasAccountData = Boolean(accountData);
  useEffect13(() => {
    closeConnectModal();
    closeAccountModal();
    closeChainModal();
  }, [hasAccountData, closeConnectModal, closeAccountModal, closeChainModal]);
  const walletConnectors = useWalletConnectors();
  const preloadImages = useCallback7(() => {
    loadImages(...walletConnectors.map((wallet) => wallet.iconUrl), ...rainbowKitChains.map((chain) => chain.iconUrl).filter(isNotNullish));
    if (!isMobile()) {
      preloadAssetsIcon();
      preloadLoginIcon();
    }
  }, [walletConnectors, rainbowKitChains]);
  useEffect13(() => {
    preloadImages();
  }, [preloadImages]);
  const displayBalance = balanceData ? `${Number(balanceData.formatted).toPrecision(3)} ${balanceData.symbol}` : void 0;
  return /* @__PURE__ */ React41.createElement(React41.Fragment, null, children({
    account: (accountData == null ? void 0 : accountData.address) ? {
      address: accountData.address,
      balanceDecimals: balanceData == null ? void 0 : balanceData.decimals,
      balanceFormatted: balanceData == null ? void 0 : balanceData.formatted,
      balanceSymbol: balanceData == null ? void 0 : balanceData.symbol,
      displayBalance,
      displayName: ensName ? formatENS(ensName) : formatAddress(accountData.address),
      ensAvatar: ensAvatar != null ? ensAvatar : void 0,
      ensName: ensName != null ? ensName : void 0,
      hasPendingTransactions
    } : void 0,
    accountModalOpen,
    chain: activeChain ? {
      hasIcon: Boolean(chainIconUrl),
      iconBackground: chainIconBackground,
      iconUrl: resolvedChainIconUrl,
      id: activeChain.id,
      name: activeChain.name,
      unsupported: activeChain.unsupported
    } : void 0,
    chainModalOpen,
    connectModalOpen,
    mounted,
    openAccountModal,
    openChainModal,
    openConnectModal
  }), /* @__PURE__ */ React41.createElement(ConnectModal, {
    onClose: closeConnectModal,
    open: connectModalOpen
  }), /* @__PURE__ */ React41.createElement(AccountModal, {
    accountData,
    balanceData,
    ensAvatar,
    ensName,
    onClose: closeAccountModal,
    onDisconnect: disconnect,
    open: accountModalOpen
  }), /* @__PURE__ */ React41.createElement(ChainModal, {
    activeChain,
    chains,
    networkError,
    onClose: closeChainModal,
    onSwitchNetwork: switchNetwork,
    open: chainModalOpen
  }));
}
ConnectButtonRenderer.displayName = "ConnectButton.Custom";

// src/components/ConnectButton/ConnectButton.tsx
var defaultProps = {
  accountStatus: "full",
  chainStatus: { largeScreen: "full", smallScreen: "icon" },
  showBalance: { largeScreen: true, smallScreen: false }
};
function ConnectButton({
  accountStatus = defaultProps.accountStatus,
  chainStatus = defaultProps.chainStatus,
  showBalance = defaultProps.showBalance
}) {
  const chains = useRainbowKitChains();
  return /* @__PURE__ */ React42.createElement(ConnectButtonRenderer, null, ({
    account,
    chain,
    mounted,
    openAccountModal,
    openChainModal,
    openConnectModal
  }) => {
    var _a, _b, _c;
    const unsupportedChain = (_a = chain == null ? void 0 : chain.unsupported) != null ? _a : false;
    return /* @__PURE__ */ React42.createElement(Box, {
      display: "flex",
      gap: "12",
      ...!mounted && {
        "aria-hidden": true,
        "style": {
          opacity: 0,
          pointerEvents: "none",
          userSelect: "none"
        }
      }
    }, mounted && account ? /* @__PURE__ */ React42.createElement(React42.Fragment, null, chain && (chains.length > 1 || unsupportedChain) && /* @__PURE__ */ React42.createElement(Box, {
      alignItems: "center",
      as: "button",
      background: unsupportedChain ? "connectButtonBackgroundError" : "connectButtonBackground",
      borderRadius: "connectButton",
      boxShadow: "connectButton",
      color: unsupportedChain ? "connectButtonTextError" : "connectButtonText",
      display: mapResponsiveValue(chainStatus, (value) => value === "none" ? "none" : "flex"),
      fontFamily: "body",
      fontWeight: "bold",
      gap: "6",
      key: unsupportedChain ? "unsupported" : "supported",
      onClick: openChainModal,
      paddingX: "10",
      paddingY: "8",
      transform: { active: "shrink", hover: "grow" },
      transition: "default",
      type: "button"
    }, unsupportedChain ? /* @__PURE__ */ React42.createElement(Box, {
      alignItems: "center",
      display: "flex",
      height: "24",
      paddingX: "4"
    }, "Wrong network") : /* @__PURE__ */ React42.createElement(Box, {
      alignItems: "center",
      display: "flex",
      gap: "6"
    }, chain.hasIcon ? /* @__PURE__ */ React42.createElement(Box, {
      display: mapResponsiveValue(chainStatus, (value) => value === "full" || value === "icon" ? "block" : "none"),
      height: "24",
      width: "24"
    }, /* @__PURE__ */ React42.createElement(AsyncImage, {
      alt: (_b = chain.name) != null ? _b : "Chain icon",
      background: chain.iconBackground,
      borderRadius: "full",
      height: "24",
      src: chain.iconUrl,
      width: "24"
    })) : null, /* @__PURE__ */ React42.createElement(Box, {
      display: mapResponsiveValue(chainStatus, (value) => {
        if (value === "icon" && !chain.iconUrl) {
          return "block";
        }
        return value === "full" || value === "name" ? "block" : "none";
      })
    }, (_c = chain.name) != null ? _c : chain.id)), /* @__PURE__ */ React42.createElement(DropdownIcon, null)), !unsupportedChain && /* @__PURE__ */ React42.createElement(Box, {
      alignItems: "center",
      as: "button",
      background: "connectButtonBackground",
      borderRadius: "connectButton",
      boxShadow: "connectButton",
      color: "connectButtonText",
      display: "flex",
      fontFamily: "body",
      fontWeight: "bold",
      onClick: openAccountModal,
      transform: { active: "shrink", hover: "grow" },
      transition: "default",
      type: "button"
    }, account.displayBalance && /* @__PURE__ */ React42.createElement(Box, {
      display: mapResponsiveValue(showBalance, (value) => value ? "block" : "none"),
      padding: "8",
      paddingLeft: "12"
    }, account.displayBalance), /* @__PURE__ */ React42.createElement(Box, {
      background: normalizeResponsiveValue(showBalance)[isMobile() ? "smallScreen" : "largeScreen"] ? "connectButtonInnerBackground" : "connectButtonBackground",
      borderColor: "connectButtonBackground",
      borderRadius: "connectButton",
      borderStyle: "solid",
      borderWidth: "2",
      color: "connectButtonText",
      fontFamily: "body",
      fontWeight: "bold",
      paddingX: "8",
      paddingY: "6",
      transition: "default"
    }, /* @__PURE__ */ React42.createElement(Box, {
      alignItems: "center",
      display: "flex",
      gap: "6",
      height: "24"
    }, /* @__PURE__ */ React42.createElement(Box, {
      display: mapResponsiveValue(accountStatus, (value) => value === "full" || value === "avatar" ? "block" : "none")
    }, /* @__PURE__ */ React42.createElement(Avatar, {
      address: account.address,
      imageUrl: account.ensAvatar,
      loading: account.hasPendingTransactions,
      size: 24
    })), /* @__PURE__ */ React42.createElement(Box, {
      alignItems: "center",
      display: "flex",
      gap: "6"
    }, /* @__PURE__ */ React42.createElement(Box, {
      display: mapResponsiveValue(accountStatus, (value) => value === "full" || value === "address" ? "block" : "none")
    }, account.displayName), /* @__PURE__ */ React42.createElement(DropdownIcon, null)))))) : /* @__PURE__ */ React42.createElement(Box, {
      background: "connectButtonBackground",
      borderRadius: "connectButton",
      transform: { active: "shrink", hover: "grow" },
      transition: "default"
    }, /* @__PURE__ */ React42.createElement(Box, {
      as: "button",
      background: "accentColor",
      borderRadius: "connectButton",
      boxShadow: "connectButton",
      color: "accentColorForeground",
      fontFamily: "body",
      fontWeight: "bold",
      onClick: openConnectModal,
      paddingX: "14",
      paddingY: "10",
      type: "button"
    }, "Connect Wallet")));
  });
}
ConnectButton.__defaultProps = defaultProps;
ConnectButton.Custom = ConnectButtonRenderer;

export {
  isAndroid,
  isIOS,
  isMobile,
  useAddress,
  useChainId,
  useTransactionStore,
  cssObjectFromTheme,
  cssStringFromTheme,
  RainbowKitProvider,
  dialogContent,
  dialogContentMobile,
  DesktopOptions,
  MobileOptions,
  ConnectButton
};
